
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>API Reference &#8212; Msprime manual</title>
    
  <link rel="stylesheet" href="_static/css/index.f658d18f9b420779cfdf24aa0a7e2d77.css">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.d3f166471bb80abb5163.js">

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Developer documentation" href="development.html" />
    <link rel="prev" title="Logging" href="logging.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="index.html">
  
  
  <h1 class="site-logo" id="site-title">Msprime manual</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Introduction
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="quickstart.html">
   Quickstart
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Using msprime
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="simulations.html">
   Running simulations
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="ancestry.html">
     Ancestry
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="demography.html">
     Demography
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mutations.html">
     Mutations
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="utilities.html">
   Utilities
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="likelihoods.html">
     Computing likelihoods
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="rate_maps.html">
     Rate Maps
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="cli.html">
     Command line interface
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="logging.html">
     Logging
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   API Reference
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="development.html">
   Developer documentation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="legacy.html">
   Legacy (version 0.x) APIs
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="changelog.html">
   Changelog
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="CITATION.html">
   Citing msprime
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   PORTING IN PROGRESS: Tutorial
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Manual for msprime __MSPRIME_VERSION__
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/api.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/tskit-dev/msprime"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/tskit-dev/msprime/issues/new?title=Issue%20on%20page%20%2Fapi.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/tskit-dev/msprime/edit/main/docs/api.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary">
   Summary
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#simulation-functions">
     Simulation functions
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ancestry-models">
     Ancestry Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mutation-models">
     Mutation Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#demography">
     Demography
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ancestry">
   Ancestry
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#models">
     Models
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mutations">
   Mutations
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Models
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#node-flags">
   Node flags
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#utilities">
   Utilities
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#rate-maps">
     Rate maps
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id2">
   Demography
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#likelihoods">
   Likelihoods
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id3">
   Mutations
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="api-reference">
<h1>API Reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h1>
<div class="section" id="summary">
<h2>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h2>
<div class="section" id="simulation-functions">
<h3>Simulation functions<a class="headerlink" href="#simulation-functions" title="Permalink to this headline">¶</a></h3>
<table class="longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sim_ancestry</span></code></a>([samples, sequence_length, …])</p></td>
<td><p>Simulates an ancestral process described by a given model, demography and samples, and return a <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> (or a sequence of replicate tree sequences).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.sim_mutations" title="msprime.sim_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sim_mutations</span></code></a>(tree_sequence[, rate, …])</p></td>
<td><p>Simulates mutations on the specified ancestry and returns the resulting <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="ancestry-models">
<h3>Ancestry Models<a class="headerlink" href="#ancestry-models" title="Permalink to this headline">¶</a></h3>
<table class="longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.StandardCoalescent" title="msprime.StandardCoalescent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">StandardCoalescent</span></code></a>()</p></td>
<td><p>The classical coalescent with recombination model (i.e., Hudson’s algorithm).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.SmcApproxCoalescent" title="msprime.SmcApproxCoalescent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SmcApproxCoalescent</span></code></a>()</p></td>
<td><p>The original SMC model defined by McVean and Cardin.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.SmcPrimeApproxCoalescent" title="msprime.SmcPrimeApproxCoalescent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SmcPrimeApproxCoalescent</span></code></a>()</p></td>
<td><p>The SMC’ model defined by Marjoram and Wall as an improvement on the original SMC.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DiscreteTimeWrightFisher</span></code></a>()</p></td>
<td><p>A discrete backwards-time Wright-Fisher model, with diploid back-and-forth recombination.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.BetaCoalescent" title="msprime.BetaCoalescent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BetaCoalescent</span></code></a>(alpha, truncation_point)</p></td>
<td><p>A Lambda-coalescent with multiple mergers in the haploid cases, or a Xi-coalescent with simultaneous multiple mergers in the polyploid case.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.DiracCoalescent" title="msprime.DiracCoalescent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DiracCoalescent</span></code></a>(psi, c)</p></td>
<td><p>A Lambda-coalescent with multiple mergers in the haploid cases, or a Xi-coalescent with simultaneous multiple mergers in the polyploid case.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.SweepGenicSelection" title="msprime.SweepGenicSelection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SweepGenicSelection</span></code></a>(position, …)</p></td>
<td><p><div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Document me</p>
</div>
</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="mutation-models">
<h3>Mutation Models<a class="headerlink" href="#mutation-models" title="Permalink to this headline">¶</a></h3>
<table class="longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.BinaryMutationModel" title="msprime.BinaryMutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BinaryMutationModel</span></code></a>()</p></td>
<td><p>Basic binary mutation model with two alleles: <code class="docutils literal notranslate"><span class="pre">&quot;0&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;1&quot;</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.JC69MutationModel" title="msprime.JC69MutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">JC69MutationModel</span></code></a>()</p></td>
<td><p>Jukes-Cantor mutation model (Jukes and Cantor 1969).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.HKYMutationModel" title="msprime.HKYMutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HKYMutationModel</span></code></a>(kappa[, …])</p></td>
<td><p>The Hasegawa, Kishino and Yano mutation model (Hasegawa et al.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.F84MutationModel" title="msprime.F84MutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">F84MutationModel</span></code></a>(kappa[, …])</p></td>
<td><p>The F84 mutation model (Felsenstein and Churchill, 1996).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.GTRMutationModel" title="msprime.GTRMutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GTRMutationModel</span></code></a>(relative_rates[, …])</p></td>
<td><p>The Generalised Time-Reversible nucleotide mutation model, a general parameterization of a time-reversible mutation process (Tavaré et al.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.BLOSUM62MutationModel" title="msprime.BLOSUM62MutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLOSUM62MutationModel</span></code></a>()</p></td>
<td><p>The BLOSUM62 model of time-reversible amino acid mutation.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.PAMMutationModel" title="msprime.PAMMutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PAMMutationModel</span></code></a>()</p></td>
<td><p>The PAM model of time-reversible amino acid mutation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.MatrixMutationModel" title="msprime.MatrixMutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MatrixMutationModel</span></code></a></p></td>
<td><p>Superclass of the specific mutation models with a finite set of states.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.InfiniteAllelesMutationModel" title="msprime.InfiniteAllelesMutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">InfiniteAllelesMutationModel</span></code></a></p></td>
<td><p>An <em>infinite alleles</em> model of mutation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.SLiMMutationModel" title="msprime.SLiMMutationModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SLiMMutationModel</span></code></a></p></td>
<td><p>An infinite-alleles model of mutation producing “SLiM-style” mutations.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="demography">
<h3>Demography<a class="headerlink" href="#demography" title="Permalink to this headline">¶</a></h3>
<table class="longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Demography</span></code></a>(populations, events, migration_matrix)</p></td>
<td><p>A description of a demographic model for an msprime simulation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a>([Ne, …])</p></td>
<td><p>A class to facilitate debugging of population parameters and migration rates in the past.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.Population" title="msprime.Population"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Population</span></code></a>(initial_size, growth_rate, name, …)</p></td>
<td><p>Define a single population in a simulation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MassMigration</span></code></a>(time, source, dest, …)</p></td>
<td><p>A mass migration event in which some fraction of the population in one deme (the <code class="docutils literal notranslate"><span class="pre">source</span></code>) simultaneously move to another deme (<code class="docutils literal notranslate"><span class="pre">dest</span></code>) during the progress of the simulation.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#msprime.PopulationParametersChange" title="msprime.PopulationParametersChange"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PopulationParametersChange</span></code></a>(time, …)</p></td>
<td><p>Changes the demographic parameters of a population at a given time.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#msprime.MigrationRateChange" title="msprime.MigrationRateChange"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MigrationRateChange</span></code></a>(time, rate, source, …)</p></td>
<td><p>Changes the rate of migration from one deme to another to a new value at a specific time.</p></td>
</tr>
</tbody>
</table>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Section for utililies like RateMap</p>
</div>
</div>
</div>
<div class="section" id="ancestry">
<h2>Ancestry<a class="headerlink" href="#ancestry" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="msprime.sim_ancestry">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">sim_ancestry</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">sequence_length</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">discrete_genome</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">recombination_rate</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_conversion_rate</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_conversion_tract_length</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">population_size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">demography</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ploidy</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">initial_state</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start_time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end_time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record_migrations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record_full_arg</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">num_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">random_seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">num_replicates</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">replicate_index</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.sim_ancestry" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates an ancestral process described by a given model, demography and
samples, and return a <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> (or a sequence of
replicate tree sequences).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> – The sampled individuals as either an integer, specifying
the number of individuals to sample in a single-population model;
or a list of <a class="reference internal" href="#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a> objects defining the properties of
groups of similar samples; or as a mapping in which the keys
are population identifiers (either an integer ID or string name)
and the values are the number of samples to take from the corresponding
population at its default sampling time. It is important to note that
samples correspond to <em>individuals</em> here, and each sampled individual
is usually associated with <span class="math notranslate nohighlight">\(k\)</span> sample <em>nodes</em> (or genomes) when
<code class="docutils literal notranslate"><span class="pre">ploidy</span></code> = <span class="math notranslate nohighlight">\(k\)</span>. See <a class="reference internal" href="ancestry.html#sec-ancestry-samples"><span class="std std-ref">Specifying samples</span></a> for further details.
Either <code class="docutils literal notranslate"><span class="pre">samples</span></code> or <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> must be specified.</p></li>
<li><p><strong>ploidy</strong> (<em>int</em>) – The number of monoploid genomes per sample individual
(Default=2). See <a class="reference internal" href="ancestry.html#sec-ancestry-ploidy"><span class="std std-ref">Ploidy</span></a> for usage examples.</p></li>
<li><p><strong>sequence_length</strong> (<em>float</em>) – The length of the genome sequence to simulate.
See <a class="reference internal" href="ancestry.html#sec-ancestry-genome-length"><span class="std std-ref">Genome length</span></a> for usage examples
for this parameter and how it interacts with other parameters.</p></li>
<li><p><strong>discrete_genome</strong> (<em>bool</em>) – If True (the default) simulation occurs
in discrete genome coordinates such that recombination and
gene conversion breakpoints always occur at integer positions.
Thus, multiple (e.g.) recombinations can occur at the same
genome position. If <code class="docutils literal notranslate"><span class="pre">discrete_genome</span></code> is False simulations
are performed using continuous genome coordinates. In this
case multiple events at precisely the same genome location are very
unlikely (but technically possible).
See <a class="reference internal" href="ancestry.html#sec-ancestry-discrete-genome"><span class="std std-ref">Discrete or continuous?</span></a> for usage examples.</p></li>
<li><p><strong>recombination_rate</strong> – The rate of recombination along the sequence;
can be either a single value (specifying a single rate over the entire
sequence) or an instance of <a class="reference internal" href="#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a>.
See <a class="reference internal" href="ancestry.html#sec-ancestry-recombination"><span class="std std-ref">Recombination</span></a> for usage examples
for this parameter and how it interacts with other parameters.</p></li>
<li><p><strong>gene_conversion_rate</strong> – The rate of gene conversion along the sequence;
can be a single value (specifying a single rate over the entire
sequence). Currently an instance of <a class="reference internal" href="#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a> is not supported.
If provided, a value for <code class="docutils literal notranslate"><span class="pre">gene_conversion_tract_length</span></code> must also be
specified. See <a class="reference internal" href="ancestry.html#sec-ancestry-gene-conversion"><span class="std std-ref">Gene conversion</span></a> for usage examples
for this parameter and how it interacts with other parameters.</p></li>
<li><p><strong>gene_conversion_tract_length</strong> – The mean length of the gene conversion
tracts. For discrete genomes the tract lengths are geometrically
distributed with mean <code class="docutils literal notranslate"><span class="pre">gene_conversion_tract_length</span></code>, which must be
greater than or equal to 1. For continuous genomes the tract lengths are
exponentially distributed with mean <code class="docutils literal notranslate"><span class="pre">gene_conversion_tract_length</span></code>,
which must be larger than 0.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em>) – The random seed. If this is not specified or <cite>None</cite>,
a high-quality random seed will be automatically generated. Valid random
seeds must be between 1 and <span class="math notranslate nohighlight">\(2^{32} - 1\)</span>.
See <a class="reference internal" href="ancestry.html#sec-ancestry-random-seed"><span class="std std-ref">Random seeds</span></a> for usage examples.</p></li>
<li><p><strong>num_replicates</strong> (<em>int</em>) – The number of replicates of the specified
parameters to simulate. If this is not specified or <cite>None</cite>,
no replication is performed and a <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> object
returned. If <code class="xref py py-obj docutils literal notranslate"><span class="pre">num_replicates</span></code> is provided, the specified
number of replicates is performed, and an iterator over the
resulting <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> objects returned.
See <a class="reference internal" href="ancestry.html#sec-ancestry-replication"><span class="std std-ref">Running replicate simulations</span></a> for examples.</p></li>
<li><p><strong>record_full_arg</strong> (<em>bool</em>) – If True, record all intermediate nodes
arising from common ancestor and recombination events in the output
tree sequence. This will result in unary nodes (i.e., nodes in marginal
trees that have only one child). Defaults to False.
See <a class="reference internal" href="ancestry.html#sec-ancestry-full-arg"><span class="std std-ref">Ancestral recombination graph</span></a> for examples.</p></li>
<li><p><strong>record_migrations</strong> (<em>bool</em>) – If True, record all migration events
that occur in the <a class="reference external" href="https://tskit.readthedocs.io/en/stable/data-model.html#sec-migration-table-definition" title="(in tskit v0.3)"><span>Migration Table</span></a> of
the output tree sequence. Defaults to False.
See <a class="reference internal" href="ancestry.html#sec-ancestry-record-migrations"><span class="std std-ref">Migration events</span></a> for examples.</p></li>
<li><p><strong>initial_state</strong> (<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><em>tskit.TreeSequence</em></a>) – If specified, initialise the
simulation from the root segments of this tree sequence and return the
completed tree sequence. Please see
<a class="reference internal" href="ancestry.html#sec-ancestry-initial-state"><span class="std std-ref">Specifying the initial state</span></a> for details of the required
properties of this tree sequence and its interactions with other parameters.
(Default: None).</p></li>
<li><p><strong>start_time</strong> (<em>float</em>) – If specified, set the initial time that the
simulation starts to this value. If not specified, the start
time is zero if performing a simulation of a set of samples,
or is the time of the oldest node if simulating from an
existing tree sequence (see the <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> parameter).
See <a class="reference internal" href="ancestry.html#sec-ancestry-start-time"><span class="std std-ref">Setting the start time</span></a> for examples.</p></li>
<li><p><strong>end_time</strong> (<em>float</em>) – If specified, terminate the simulation at the
specified time. In the returned tree sequence, all rootward paths from
samples with time &lt; <code class="docutils literal notranslate"><span class="pre">end_time</span></code> will end in a node with one child with
time equal to end_time. Any sample nodes with time &gt;= <code class="docutils literal notranslate"><span class="pre">end_time</span></code> will
also be present in the output tree sequence. If not specified or <code class="docutils literal notranslate"><span class="pre">None</span></code>,
run the simulation until all samples have an MRCA at all positions in
the genome. See <a class="reference internal" href="ancestry.html#sec-ancestry-end-time"><span class="std std-ref">Stopping simulations early</span></a> for examples.</p></li>
<li><p><strong>model</strong> (<em>str</em><em> or </em><em>simulation model instance</em>) – The ancestry model to use.
This can either be a string (e.g., <code class="docutils literal notranslate"><span class="pre">&quot;smc_prime&quot;</span></code>) or an instance of
a simulation model class (e.g, <code class="docutils literal notranslate"><span class="pre">msprime.DiscreteTimeWrightFisher()</span></code>.
Please see the <a class="reference internal" href="ancestry.html#sec-ancestry-models"><span class="std std-ref">Models</span></a> section for more details
on specifying simulations models.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> object representing the results
of the simulation if no replication is performed, or an
iterator over the independent replicates simulated if the
<code class="xref py py-obj docutils literal notranslate"><span class="pre">num_replicates</span></code> parameter has been used.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> or an iterator over
<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> replicates.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.SampleSet">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">SampleSet</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_samples</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">population</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>None<span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">time</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ploidy</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>None<span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.SampleSet" title="Permalink to this definition">¶</a></dt>
<dd><p>TODO document</p>
</dd></dl>

<dl class="py class">
<dt id="msprime.AncestryModelChange">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">AncestryModelChange</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>Union<span class="p">[</span>str<span class="p">, </span>msprime.ancestry.AncestryModel<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.AncestryModelChange" title="Permalink to this definition">¶</a></dt>
<dd><p>An event representing a change of underlying <a class="reference internal" href="ancestry.html#sec-ancestry-models"><span class="std std-ref">ancestry model</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<em>float</em>) – The time at which the ancestry model changes
to the new model, in generations. After this time, all internal
tree nodes, edges and migrations are the result of the new model.
If time is set to None (the default), the model change will occur
immediately after the previous model has completed. If time is a
callable, the time at which the model changes is the result
of calling this function with the time that the previous model
started with as a parameter.</p></li>
<li><p><strong>model</strong> (<em>str</em><em> or </em><em>ancestry model instance</em>) – The new simulation model to use.
This can either be a string (e.g., <code class="docutils literal notranslate"><span class="pre">&quot;smc_prime&quot;</span></code>) or an instance of
an ancestry model class (e.g, <code class="docutils literal notranslate"><span class="pre">msprime.DiscreteTimeWrightFisher()</span></code>.
Please see the <a class="reference internal" href="ancestry.html#sec-ancestry-models"><span class="std std-ref">Models</span></a> section for more details
on specifying these models. If this is None (the default) the model is
changed to the standard coalescent.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<div class="section" id="models">
<h3>Models<a class="headerlink" href="#models" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt id="msprime.StandardCoalescent">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">StandardCoalescent</code><a class="headerlink" href="#msprime.StandardCoalescent" title="Permalink to this definition">¶</a></dt>
<dd><p>The classical coalescent with recombination model (i.e., Hudson’s algorithm).
The string <code class="docutils literal notranslate"><span class="pre">&quot;hudson&quot;</span></code> can be used to refer to this model.</p>
<p>This is the default simulation model.</p>
</dd></dl>

<dl class="py class">
<dt id="msprime.SmcApproxCoalescent">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">SmcApproxCoalescent</code><a class="headerlink" href="#msprime.SmcApproxCoalescent" title="Permalink to this definition">¶</a></dt>
<dd><p>The original SMC model defined by McVean and Cardin. This
model is implemented using a naive rejection sampling approach
and so it may not be any more efficient to simulate than the
standard Hudson model.</p>
<p>The string <code class="docutils literal notranslate"><span class="pre">&quot;smc&quot;</span></code> can be used to refer to this model.</p>
</dd></dl>

<dl class="py class">
<dt id="msprime.SmcPrimeApproxCoalescent">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">SmcPrimeApproxCoalescent</code><a class="headerlink" href="#msprime.SmcPrimeApproxCoalescent" title="Permalink to this definition">¶</a></dt>
<dd><p>The SMC’ model defined by Marjoram and Wall as an improvement on the
original SMC. model is implemented using a naive rejection sampling
approach and so it may not be any more efficient to simulate than the
standard Hudson model.</p>
<p>The string <code class="docutils literal notranslate"><span class="pre">&quot;smc_prime&quot;</span></code> can be used to refer to this model.</p>
</dd></dl>

<dl class="py class">
<dt id="msprime.DiscreteTimeWrightFisher">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">DiscreteTimeWrightFisher</code><a class="headerlink" href="#msprime.DiscreteTimeWrightFisher" title="Permalink to this definition">¶</a></dt>
<dd><p>A discrete backwards-time Wright-Fisher model, with diploid back-and-forth
recombination. The string <code class="docutils literal notranslate"><span class="pre">&quot;dtwf&quot;</span></code> can be used to refer to this model.</p>
<p>Wright-Fisher simulations are performed very similarly to coalescent
simulations, with all parameters denoting the same quantities in both
models. Because events occur at discrete times however, the order in which
they occur matters. Each generation consists of the following ordered
events:</p>
<ul class="simple">
<li><p>Migration events. As in the Hudson coalescent, these move single extant
lineages between populations. Because migration events occur before
lineages choose parents, migrant lineages choose parents from their new
population in the same generation.</p></li>
<li><p>Demographic events. All events with <cite>previous_generation &lt; event_time &lt;=
current_generation</cite> are carried out here.</p></li>
<li><p>Lineages draw parents. Each (monoploid) extant lineage draws a parent
from their current population.</p></li>
<li><p>Diploid recombination. Each parent is diploid, so all child lineages
recombine back-and-forth into the same two parental genome copies. These
become two independent lineages in the next generation.</p></li>
<li><p>Historical sampling events. All historical samples with
<cite>previous_generation &lt; sample_time &lt;= current_generation</cite> are inserted.</p></li>
</ul>
</dd></dl>

<dl class="py class">
<dt id="msprime.BetaCoalescent">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">BetaCoalescent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">truncation_point</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">1.7976931348623157e+308</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.BetaCoalescent" title="Permalink to this definition">¶</a></dt>
<dd><p>A Lambda-coalescent with multiple mergers in the haploid cases, or a
Xi-coalescent with simultaneous multiple mergers in the polyploid case.</p>
<p>There are two main differences between the Beta-coalescent and the
standard coalescent. Firstly, the number of lineages that take part in each
common ancestor event is random, with distribution determined by moments of
the <span class="math notranslate nohighlight">\(Beta(2 - \alpha, \alpha)\)</span>-distribution. In particular, when there
are <span class="math notranslate nohighlight">\(n\)</span> lineages, each set of <span class="math notranslate nohighlight">\(k \leq n\)</span> of them participates in a
common ancestor event at rate</p>
<div class="math notranslate nohighlight">
\[\frac{1}{B(2 - \alpha, \alpha)}
\int_0^1 x^{k - \alpha - 1} (1 - x)^{n - k + \alpha - 1} dx,\]</div>
<p>where <span class="math notranslate nohighlight">\(B(2 - \alpha, \alpha)\)</span> is the Beta-function.</p>
<p>If ploidy = 1, then all participating lineages merge into one common ancestor,
corresponding to haploid, single-parent reproduction.
If ploidy = <span class="math notranslate nohighlight">\(p &gt; 1\)</span>, all participating lineages split randomly into
<span class="math notranslate nohighlight">\(2 p\)</span> groups, corresponding to two-parent reproduction with <span class="math notranslate nohighlight">\(p\)</span> copies
of each chromosome per parent. All lineages within each group merge simultaneously.</p>
<p>Secondly, the number of generations between common ancestor events predicted by the
Beta-coalescent is proportional to <span class="math notranslate nohighlight">\(N^{\alpha - 1}\)</span>, where <span class="math notranslate nohighlight">\(N\)</span> is
the population size. Specifically, the mean number of generations until
two lineages undergo a common ancestor event is</p>
<div class="math notranslate nohighlight">
\[G = \frac{m^{\alpha} N^{\alpha - 1}}{\alpha B(2 - \alpha, \alpha)},\]</div>
<p>if ploidy = 1, and</p>
<div class="math notranslate nohighlight">
\[G = \frac{2 p m^{\alpha} (N / 2)^{\alpha - 1}}
    {\alpha B(2 - \alpha, \alpha)},\]</div>
<p>if ploidy = <span class="math notranslate nohighlight">\(p &gt; 1\)</span>, where <span class="math notranslate nohighlight">\(m\)</span> is the mean number of juveniles per
family given by</p>
<div class="math notranslate nohighlight">
\[m = 2 + \frac{2^{\alpha}}{3^{\alpha - 1} (\alpha - 1)},\]</div>
<p>if ploidy &gt; 1, and</p>
<div class="math notranslate nohighlight">
\[m = 1 + \frac{1}{2^{\alpha - 1} (\alpha - 1)},\]</div>
<p>if ploidy = 1.</p>
<p>In the polyploid case we divide the population size <span class="math notranslate nohighlight">\(N\)</span> by two
because we assume the <span class="math notranslate nohighlight">\(N\)</span> polyploid individuals form <span class="math notranslate nohighlight">\(N / 2\)</span>
two-parent families in which reproduction takes place.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The number of generations between common ancestor events <span class="math notranslate nohighlight">\(G\)</span> depends
both on the population size <span class="math notranslate nohighlight">\(N\)</span> and <span class="math notranslate nohighlight">\(\alpha\)</span>,
and can be dramatically shorter than in the case of the
standard coalescent. For <span class="math notranslate nohighlight">\(\alpha \approx 1\)</span> that is due to
insensitivity of <span class="math notranslate nohighlight">\(G\)</span> to <span class="math notranslate nohighlight">\(N\)</span> — see
<a class="reference internal" href="ancestry.html#sec-ancestry-models-multiple-mergers"><span class="std std-ref">Multiple merger coalescents</span></a> for an illustration.
For <span class="math notranslate nohighlight">\(\alpha \approx 2\)</span>, <span class="math notranslate nohighlight">\(G\)</span> is almost linear in
<span class="math notranslate nohighlight">\(N\)</span>, but can nevertheless be small because
<span class="math notranslate nohighlight">\(B(2 - \alpha, \alpha) \rightarrow \infty\)</span> as
<span class="math notranslate nohighlight">\(\alpha \rightarrow 2\)</span>. As a result, population sizes
must often be many orders of magnitude larger than census population sizes
to obtain realistic amounts of diversity in simulated samples.</p>
</div>
<p>See <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0304414903000280">Schweinsberg (2003)</a>
for the derivation of the common ancestor event rate,
as well as the number of generations between common ancestor events.
Note however that Schweinsberg (2003) only covers the haploid case.
For details of the diploid extension, see
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3527250/">Blath et al. (2013)</a>,
and <a class="reference external" href="https://projecteuclid.org/euclid.ejp/1527818427">Birkner et al. (2018)</a>
for a diploid version of the Schweinsberg (2003) model specifically.
The general polyploid model is analogous to the diploid case, with
<span class="math notranslate nohighlight">\(2 p\)</span> available copies of parental chromsomes per common ancestor event,
and hence up to <span class="math notranslate nohighlight">\(2 p\)</span> simultaneous mergers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alpha</strong> (<em>float</em>) – Determines the degree of skewness in the family size
distribution, and must satisfy <span class="math notranslate nohighlight">\(1 &lt; \alpha &lt; 2\)</span>. Smaller values of
<span class="math notranslate nohighlight">\(\alpha\)</span> correspond to greater skewness, and <span class="math notranslate nohighlight">\(\alpha = 2\)</span>
would coincide with the standard coalescent.</p></li>
<li><p><strong>truncation_point</strong> (<em>float</em>) – The maximum number of juveniles <span class="math notranslate nohighlight">\(K\)</span> born to
one family as a fraction of the population size <span class="math notranslate nohighlight">\(N\)</span>. Must satisfy
<span class="math notranslate nohighlight">\(0 &lt; K \leq \inf\)</span>. Determines the maximum fraction of the population
replaced by offspring in one reproduction event, <span class="math notranslate nohighlight">\(\tau\)</span>, via
<span class="math notranslate nohighlight">\(\tau = K / (K + m)\)</span>, where <span class="math notranslate nohighlight">\(m\)</span> is the mean juvenile number
above. The default is <span class="math notranslate nohighlight">\(K = \inf\)</span>, which corresponds to the standard
Beta-coalescent with <span class="math notranslate nohighlight">\(\tau = 1\)</span>. When <span class="math notranslate nohighlight">\(K &lt; \inf\)</span>, the number of
lineages participating in a common ancestor event is determined by moments
of the Beta:math:<cite>(2 - alpha, alpha)</cite> distribution conditioned on not
exceeding <span class="math notranslate nohighlight">\(\tau\)</span>, and the Beta-function in the expression
for <span class="math notranslate nohighlight">\(G\)</span> is replaced by the incomplete Beta-function
<span class="math notranslate nohighlight">\(B(\tau; 2 - \alpha, \alpha)\)</span>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.DiracCoalescent">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">DiracCoalescent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">psi</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">c</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.DiracCoalescent" title="Permalink to this definition">¶</a></dt>
<dd><p>A Lambda-coalescent with multiple mergers in the haploid cases, or a
Xi-coalescent with simultaneous multiple mergers in the polyploid case.</p>
<p>The Dirac-coalescent is an implementation of the model of
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3527250/">Blath et al. (2013)</a>
The simulation proceeds similarly to the standard coalescent.
In addition to binary common ancestor events at rate <span class="math notranslate nohighlight">\(n (n - 1) / 2\)</span> when
there are <span class="math notranslate nohighlight">\(n\)</span> lineages, potential multiple merger events take place
at rate <span class="math notranslate nohighlight">\(c &gt; 0\)</span>. Each lineage participates in each multiple merger
event independently with probability <span class="math notranslate nohighlight">\(0 &lt; \psi \leq 1\)</span>.</p>
<p>If ploidy = 1, then all participating lineages merge into one common ancestor,
corresponding to haploid, single-parent reproduction.
If ploidy = <span class="math notranslate nohighlight">\(p &gt; 1\)</span>, all participating lineages split randomly into
<span class="math notranslate nohighlight">\(2 p\)</span> groups, corresponding to two-parent reproduction with <span class="math notranslate nohighlight">\(p\)</span> copies
of each chromosome per parent. All lineages within each group merge simultaneously.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The Dirac-coalescent is obtained as a scaling limit of Moran models,
rather than Wright-Fisher models. As a consequence, the number of generations
between coalescence events is proportional to <span class="math notranslate nohighlight">\(N^2\)</span>,
rather than <span class="math notranslate nohighlight">\(N\)</span> generations as in the standard coalescent.
See <a class="reference internal" href="ancestry.html#sec-ancestry-models-multiple-mergers"><span class="std std-ref">Multiple merger coalescents</span></a> for an illustration
of how this affects simulation output in practice.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c</strong> (<em>float</em>) – Determines the rate of potential multiple merger events.
We require <span class="math notranslate nohighlight">\(c &gt; 0\)</span>.</p></li>
<li><p><strong>psi</strong> (<em>float</em>) – Determines the fraction of the population replaced by
offspring in one large reproduction event, i.e. one reproduction event
giving rise to potential multiple mergers when viewed backwards in time.
We require <span class="math notranslate nohighlight">\(0 &lt; \psi \leq 1\)</span>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.SweepGenicSelection">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">SweepGenicSelection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start_frequency</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end_frequency</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dt</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.SweepGenicSelection" title="Permalink to this definition">¶</a></dt>
<dd><div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Document me</p>
</div>
</dd></dl>

</div>
</div>
<div class="section" id="mutations">
<h2>Mutations<a class="headerlink" href="#mutations" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>Models<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt id="msprime.BinaryMutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">BinaryMutationModel</code><a class="headerlink" href="#msprime.BinaryMutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Basic binary mutation model with two alleles: <code class="docutils literal notranslate"><span class="pre">&quot;0&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;1&quot;</span></code>. The ancestral
allele is always <code class="docutils literal notranslate"><span class="pre">&quot;0&quot;</span></code> and the transition probabilities are set such that each
mutation always transitions to the other state.</p>
<p>This is a <a class="reference internal" href="#msprime.MatrixMutationModel" title="msprime.MatrixMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixMutationModel</span></code></a> with alleles <code class="docutils literal notranslate"><span class="pre">[&quot;0&quot;,</span> <span class="pre">&quot;1&quot;]</span></code>,
root distribution <code class="docutils literal notranslate"><span class="pre">[1.0,</span> <span class="pre">0.0]</span></code>, and transition matrix
<code class="docutils literal notranslate"><span class="pre">[[0.0,</span> <span class="pre">1.0],</span> <span class="pre">[1.0,</span> <span class="pre">0.0]]</span></code>.</p>
</dd></dl>

<dl class="py class">
<dt id="msprime.JC69MutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">JC69MutationModel</code><a class="headerlink" href="#msprime.JC69MutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Jukes-Cantor mutation model (Jukes and Cantor 1969). Based on the standard ACGT
nucleotides as alleleic states, this model assumes equal probabilities for
ancestral state and equal probabilities for transitions to other alleles.</p>
<p>This is a <a class="reference internal" href="#msprime.MatrixMutationModel" title="msprime.MatrixMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixMutationModel</span></code></a> with alleles <code class="docutils literal notranslate"><span class="pre">[&quot;A&quot;,</span> <span class="pre">&quot;C&quot;,</span> <span class="pre">&quot;G&quot;,</span> <span class="pre">&quot;T&quot;]</span></code>,
root distribution <code class="docutils literal notranslate"><span class="pre">[1/4,</span> <span class="pre">1/4,</span> <span class="pre">1/4,</span> <span class="pre">1/4]</span></code>, and transition matrix
<code class="docutils literal notranslate"><span class="pre">[[0,</span> <span class="pre">1/3,</span> <span class="pre">1/3,</span> <span class="pre">1/3],</span> <span class="pre">[1/3,</span> <span class="pre">0,</span> <span class="pre">1/3,</span> <span class="pre">1/3],</span> <span class="pre">[1/3,</span> <span class="pre">1/3,</span> <span class="pre">0,</span> <span class="pre">1/3],</span> <span class="pre">[1/3,</span> <span class="pre">1/3,</span> <span class="pre">1/3,</span> <span class="pre">0]]</span></code>.
It has no free parameters.</p>
<p>Citation: <em>Jukes TH, Cantor CR (1969). Evolution of Protein Molecules. pp. 21-132.</em></p>
</dd></dl>

<dl class="py class">
<dt id="msprime.HKYMutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">HKYMutationModel</code><a class="headerlink" href="#msprime.HKYMutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>The Hasegawa, Kishino and Yano mutation model (Hasegawa et al. 1985). Based on the
standard ACGT nucleotides as alleleic states, this model allows different rates for
transitions and transversions, and sets an equilibrium frequency for each nucleotide.
In addition a custom ancestral frequency (<code class="docutils literal notranslate"><span class="pre">root_distribution</span></code>) can be specified.
With <code class="docutils literal notranslate"><span class="pre">kappa=1.0</span></code> and the default values of the other arguments this model is equal
to <a class="reference internal" href="#msprime.JC69MutationModel" title="msprime.JC69MutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">JC69MutationModel</span></code></a>. This model is similar to <a class="reference internal" href="#msprime.F84MutationModel" title="msprime.F84MutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">F84MutationModel</span></code></a>
but with a differing parametrisation for <code class="docutils literal notranslate"><span class="pre">kappa</span></code>.</p>
<p>This model is parameterized by <span class="math notranslate nohighlight">\(\kappa\)</span> (<code class="docutils literal notranslate"><span class="pre">kappa</span></code>), the ratio of
transition to transversion mutation rates, and <span class="math notranslate nohighlight">\(\pi\)</span>
(<code class="docutils literal notranslate"><span class="pre">equilibrium_frequencies</span></code>), the vector of equilibrium nucleotide
frequencies. If this mutation model is used with a <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> of
<span class="math notranslate nohighlight">\(\mu\)</span>, then the mutation rate from the i-th to the j-th allele is
equal to <span class="math notranslate nohighlight">\(Q_{ij}\)</span>, where</p>
<p><span class="math notranslate nohighlight">\(\mathbf{Q} =
\frac{\mu}{M} \times
\begin{bmatrix}
\cdot        &amp;  \pi_C         &amp; \kappa \pi_G &amp; \pi_T           \\
\pi_A        &amp;  \cdot         &amp; \pi_G        &amp; \kappa \pi_T    \\
\kappa \pi_A &amp;  \pi_C         &amp; \cdot        &amp; \pi_T           \\
\pi_A        &amp;  \kappa \pi_C  &amp; \pi_G        &amp; \cdot
\end{bmatrix}\)</span></p>
<p>Here <span class="math notranslate nohighlight">\(M\)</span> is an overall scaling factor on the mutation rate that can
be computed as follows: let <span class="math notranslate nohighlight">\(q_{i,j} = \pi_j\)</span> if i&lt;-&gt;j is a transversion,
and <span class="math notranslate nohighlight">\(q_{ij} = \kappa \pi_j\)</span> otherwise. Set <span class="math notranslate nohighlight">\(q_{ii} = 0\)</span>, and
then let <span class="math notranslate nohighlight">\(M\)</span> be the largest row sum of <span class="math notranslate nohighlight">\(q\)</span>, i.e.,
<span class="math notranslate nohighlight">\(M = \max_i \left( \sum_{j \neq i} q_{ij} \right)\)</span>.
Then, this implementation as a <a class="reference internal" href="#msprime.MatrixMutationModel" title="msprime.MatrixMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixMutationModel</span></code></a> has the transition
matrix <span class="math notranslate nohighlight">\(P_{ij} = q_{ij} / M\)</span>, and <span class="math notranslate nohighlight">\(P_{ii}\)</span> chosen so rows sum to one.</p>
<p>Note also that <span class="math notranslate nohighlight">\(\kappa\)</span> is the ratio of <em>individual</em> mutation rates,
not the ratio of <em>total</em> transition to transversion rates, which would
be <span class="math notranslate nohighlight">\(\kappa/2\)</span>, as is used in some parameterizations.</p>
<p>Citation: <em>Hasegawa M, Kishino H, Yano T (1985).  “Dating of the human-ape
splitting by a molecular clock of mitochondrial DNA”. Journal of Molecular
Evolution. 22 (2): 160–74.</em></p>
<p>Note that this implementation has the root distribution as a separate
parameter, although it defaults to the equilibrium distribution. If you
set the root distribution to be different than the equilbrium distribution,
then you have a nonequilibrium model, and you should make sure that’s what
you want.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa</strong> (<em>float</em>) – Scaling parameter by which the transition matrix is modified for
transitions (A&lt;&gt;G, C&lt;&gt;T). For example a value of <code class="docutils literal notranslate"><span class="pre">0</span></code> means that no transitions
will occur, <code class="docutils literal notranslate"><span class="pre">1</span></code> makes the probability of transitions and
transversions equal.</p></li>
<li><p><strong>equilibrium_frequencies</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – An array of float values of length
4 which should sum to 1. These values are used to determine equilibrium
frequencies (long-term averages) of the alleles. Defaults to
[0.25, 0.25, 0.25, 0.25], i.e., all nucleotides equally likely.</p></li>
<li><p><strong>root_distribution</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – An array of float values of length 4
which should sum to 1. These values are used to determine the ancestral state of
each mutational site. Defaults to the value of <code class="docutils literal notranslate"><span class="pre">equilibrium_frequencies</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.F84MutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">F84MutationModel</code><a class="headerlink" href="#msprime.F84MutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>The F84 mutation model (Felsenstein and Churchill, 1996). Based on the
standard ACGT nucleotides as alleleic states, this model takes into account
transitions and transversions, and sets an equilibrium frequency for each nucleotide.
In addition a custom ancestral frequency (<code class="docutils literal notranslate"><span class="pre">root_distribution</span></code>) can be specified.
With <code class="docutils literal notranslate"><span class="pre">kappa=1.0</span></code> and the default values of the other arguments this model is equal
to <a class="reference internal" href="#msprime.JC69MutationModel" title="msprime.JC69MutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">JC69MutationModel</span></code></a>. This model is similar to <a class="reference internal" href="#msprime.HKYMutationModel" title="msprime.HKYMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">HKYMutationModel</span></code></a>
but with a differing parametrisation for <code class="docutils literal notranslate"><span class="pre">kappa</span></code>.</p>
<p>This model is parameterized by <span class="math notranslate nohighlight">\(\kappa\)</span> (<code class="docutils literal notranslate"><span class="pre">kappa</span></code>), the ratio of
transition to transversion mutation rates, and <span class="math notranslate nohighlight">\(\pi\)</span>
(<code class="docutils literal notranslate"><span class="pre">equilibrium_frequencies</span></code>), the vector of equilibrium nucleotide
frequencies. If this mutation model is used with a <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> of
<span class="math notranslate nohighlight">\(\mu\)</span>, then the mutation rate from the i-th to the j-th allele is
equal to <span class="math notranslate nohighlight">\(Q_{ij}\)</span>, where</p>
<p><span class="math notranslate nohighlight">\(\mathbf{Q} =
\frac{\mu}{M} \times
\begin{bmatrix}
\cdot &amp; \pi_C
&amp; \left(1 + \frac{\kappa-1}{\pi_A + \pi_G}\right) \pi_G &amp; \pi_T \\
\pi_A &amp; \cdot &amp; \pi_G
&amp; \left(1 + \frac{\kappa-1}{\pi_C + \pi_T}\right) \pi_T \\
\left(1 + \frac{\kappa-1}{\pi_A + \pi_G}\right) \pi_A &amp; \pi_C
&amp; \cdot &amp; \pi_T \\
\pi_A &amp; \left(1 + \frac{\kappa-1}{\pi_C + \pi_T}\right) \pi_C
&amp; \pi_G &amp; \cdot
\end{bmatrix}\)</span></p>
<p>Here <span class="math notranslate nohighlight">\(M\)</span> is an overall scaling factor on the mutation rate that can be
computed as follows: let <span class="math notranslate nohighlight">\(q_{ij} = \pi_j\)</span> if i&lt;-&gt;j is a transversion,
<span class="math notranslate nohighlight">\(q_{ij} = (1 + (\kappa-1)/(\pi_A + \pi_G)) \pi_j\)</span> if i and j are A and
G in some order, and <span class="math notranslate nohighlight">\(q_{ij} = (1 + (\kappa-1)/(\pi_C + \pi_T)) \pi_j\)</span>
if i and j are C and T in some order. Set <span class="math notranslate nohighlight">\(q_{ii} = 0\)</span>, and then let
<span class="math notranslate nohighlight">\(M\)</span> be the largest row sum of <span class="math notranslate nohighlight">\(q\)</span>, i.e.,
<span class="math notranslate nohighlight">\(M = \max_i \left( \sum_{j \neq i} q_{ij} \right)\)</span>.
Then, this implementation as a <a class="reference internal" href="#msprime.MatrixMutationModel" title="msprime.MatrixMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixMutationModel</span></code></a> has the
transition matrix <span class="math notranslate nohighlight">\(P_{ij} = q_{ij} / M\)</span>, and <span class="math notranslate nohighlight">\(P_{ii}\)</span> chosen so
rows sum to one.</p>
<p>Citation: <em>Felsenstein J, Churchill GA (January 1996). “A Hidden Markov
Model approach to variation among sites in rate of evolution”. Molecular
Biology and Evolution. 13 (1): 93–104.</em></p>
<p>Note that this implementation has the root distribution as a separate
parameter, although it defaults to the equilibrium distribution. If you
set the root distribution to be different than the equilbrium distribution,
then you have a nonequilibrium model, and you should make sure that’s what
you want.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa</strong> (<em>float</em>) – Scaling parameter by which the transition matrix is modified for
transitions (A&lt;&gt;G, C&lt;&gt;T). Must be greater than or equal to the larger of
<span class="math notranslate nohighlight">\((\pi_A+\pi_G)/(\pi_C+\pi_T)\)</span>
and <span class="math notranslate nohighlight">\((\pi_C+\pi_T)/(\pi_A+\pi_G)\)</span>.</p></li>
<li><p><strong>equilibrium_frequencies</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – An array of float values of length
4 which should sum to 1. These values are used to determine equilibrium
frequencies (long-term averages) of the alleles. Defaults to
[0.25, 0.25, 0.25, 0.25], i.e all nucleotides equally likely.</p></li>
<li><p><strong>root_distribution</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – An array of float values of length 4
which should sum to 1. These values are used to determine the ancestral state of
each mutational site. Defaults to the value of <code class="docutils literal notranslate"><span class="pre">equilibrium_frequencies</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.GTRMutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">GTRMutationModel</code><a class="headerlink" href="#msprime.GTRMutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>The Generalised Time-Reversible nucleotide mutation model, a general
parameterization of a time-reversible mutation process (Tavaré et al.
1986). It allows specification of per-nucleotide equilibrium frequencies
and equilibrium transition rates.</p>
<p>This model is parameterized by the vector <span class="math notranslate nohighlight">\(r\)</span> (<code class="docutils literal notranslate"><span class="pre">relative_rates</span></code>),
and <span class="math notranslate nohighlight">\(\pi\)</span> (<code class="docutils literal notranslate"><span class="pre">equilibrium_frequencies</span></code>), the vector of equilibrium nucleotide
frequencies. The entries of <code class="docutils literal notranslate"><span class="pre">relative_rates</span></code> are, in this order,
<span class="math notranslate nohighlight">\((r_{AC}, r_{AG}, r_{AT}, r_{CG}, r_{CT}, r_{GT})\)</span>. If this mutation model
is used with a <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> of <span class="math notranslate nohighlight">\(\mu\)</span>, then the mutation rate from the
i-th to the j-th allele is <span class="math notranslate nohighlight">\(Q_{ij}\)</span>, where</p>
<p><span class="math notranslate nohighlight">\(\mathbf{Q} =
\frac{\mu}{M} \times
\begin{bmatrix}
\cdot        &amp;  r_{AC} \pi_C  &amp; r_{AG} \pi_G &amp; r_{AT} \pi_T    \\
r_{AC} \pi_A &amp;  \cdot         &amp; r_{CG} \pi_G &amp; r_{CT} \pi_T    \\
r_{AG} \pi_A &amp;  r_{CG} \pi_C  &amp; \cdot        &amp; r_{GT} \pi_T    \\
r_{AT} \pi_A &amp;  r_{CT} \pi_C  &amp; r_{GT} \pi_G &amp; \cdot
\end{bmatrix}\)</span></p>
<p>Here <span class="math notranslate nohighlight">\(M\)</span> is an overall scaling factor on the mutation rate that can be
computed as follows: let <span class="math notranslate nohighlight">\(q_{ij} = r_{ij} \pi_j\)</span>,
and <span class="math notranslate nohighlight">\(q_{ii} = 0\)</span>, and then let <span class="math notranslate nohighlight">\(M\)</span> be the largest row sum of
<span class="math notranslate nohighlight">\(q\)</span>, i.e., <span class="math notranslate nohighlight">\(M = \max_i \left( \sum_{j \neq i} q_{ij}
\right)\)</span>. Then, this implementation as a <a class="reference internal" href="#msprime.MatrixMutationModel" title="msprime.MatrixMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixMutationModel</span></code></a> has
the transition matrix <span class="math notranslate nohighlight">\(P_{ij} = q_{ij} / M\)</span>, and <span class="math notranslate nohighlight">\(P_{ii}\)</span>
chosen so rows sum to one.</p>
<p>Citation: <em>Tavaré S (1986). “Some Probabilistic and Statistical Problems in the
Analysis of DNA Sequences”. Lectures on Mathematics in the Life Sciences. 17:
57–86.</em></p>
<p>Note that this implementation has the root distribution as a separate
parameter, although it defaults to the equilibrium distribution. If you
set the root distribution to be different than the equilbrium distribution,
then you have a nonequilibrium model, and you should make sure that’s what
you want.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relative_rates</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – Controls the relative rates of
mutation between each pair of nucleotides, in order “A&lt;&gt;C”, “A&lt;&gt;G”, “A&lt;&gt;T”,
“C&lt;&gt;G”, “C&lt;&gt;T”, and “G&lt;&gt;T”.</p></li>
<li><p><strong>equilibrium_frequencies</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – The equilibrium base
frequencies, a list of four probabilities that sum to one. (Default: equal
frequencies.)</p></li>
<li><p><strong>root_distribution</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – An array of float values of length 4
which should sum to 1. These values are used to determine the ancestral state of
each mutational site. Defaults to the value of <code class="docutils literal notranslate"><span class="pre">equilibrium_frequencies</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.BLOSUM62MutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">BLOSUM62MutationModel</code><a class="headerlink" href="#msprime.BLOSUM62MutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>The BLOSUM62 model of time-reversible amino acid mutation. This model has
no free parameters.</p>
<p>The model is parameterized by a 20-by-20 symmetric matrix of relative rates,
<span class="math notranslate nohighlight">\(B\)</span>, and a vector of amino acid equilibrium frequencies, <span class="math notranslate nohighlight">\(\pi\)</span>
(for the precise order, see this model’s <code class="xref py py-attr docutils literal notranslate"><span class="pre">alleles</span></code> attribute).
If this mutation model is used with a <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> of <span class="math notranslate nohighlight">\(\mu\)</span>, then
the mutation rate from the i-th to the j-th allele is <span class="math notranslate nohighlight">\(Q_{ij}\)</span>, where</p>
<p><span class="math notranslate nohighlight">\(Q_{ij} = \frac{\mu}{M} B_{ij} \pi_j,\)</span></p>
<p>where <span class="math notranslate nohighlight">\(M\)</span> is an overall scaling factor chosen so that the largest
row sum in the matrix <span class="math notranslate nohighlight">\(B_{ij} \pi_j /M\)</span> is equal to one.
In this model, <span class="math notranslate nohighlight">\(M = 1.726203705809619\)</span>.
This implementation as a <a class="reference internal" href="#msprime.MatrixMutationModel" title="msprime.MatrixMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixMutationModel</span></code></a> has transition matrix
<span class="math notranslate nohighlight">\(P_{ij} = B_{ij} \pi_{j} / M\)</span>, and <span class="math notranslate nohighlight">\(P_{ii}\)</span> chosen so rows sum to one,
and root distribution equal to the equilibrium frequencies, <span class="math notranslate nohighlight">\(\pi\)</span>,
so the matrix <span class="math notranslate nohighlight">\(B\)</span> can be recovered as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">BLOSUM62</span><span class="p">()</span>
<span class="n">M</span> <span class="o">=</span> <span class="mf">1.726203705809619</span>
<span class="n">B</span> <span class="o">=</span> <span class="n">M</span> <span class="o">*</span> <span class="n">model</span><span class="o">.</span><span class="n">transition_matrix</span> <span class="o">/</span> <span class="n">model</span><span class="o">.</span><span class="n">root_distribution</span>
</pre></div>
</div>
<p>Citation: The values of <span class="math notranslate nohighlight">\(B\)</span> and <span class="math notranslate nohighlight">\(\pi\)</span> used here were copied from
<a class="reference external" href="http://tree.bio.ed.ac.uk/software/seqgen/">Seq-Gen</a>.
The original citation is: <em>Henikoff, S., and J. G. Henikoff. 1992. PNAS USA
89:10915-10919,</em> and the numerical values were provided in: <em>Yu,Y.-K.,
Wootton,J.C. and Altschul,S.F. (2003) The compositional adjustment of amino
acid substitution matrices. Proc. Natl Acad. Sci., USA, 100, 15688–15693.</em></p>
</dd></dl>

<dl class="py class">
<dt id="msprime.PAMMutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">PAMMutationModel</code><a class="headerlink" href="#msprime.PAMMutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>The PAM model of time-reversible amino acid mutation. This model has no
free parameters.</p>
<p>The model is parameterized by a 20-by-20 symmetric matrix of relative rates,
<span class="math notranslate nohighlight">\(B\)</span>, and a vector of amino acid equilibrium frequencies, <span class="math notranslate nohighlight">\(\pi\)</span>
(for the precise order, see this model’s <code class="xref py py-attr docutils literal notranslate"><span class="pre">alleles</span></code> attribute).
If this mutation model is used with a <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> of <span class="math notranslate nohighlight">\(\mu\)</span>, then
the mutation rate from the i-th to the j-th allele is <span class="math notranslate nohighlight">\(Q_{ij}\)</span>, where</p>
<p><span class="math notranslate nohighlight">\(Q_{ij} = \frac{\mu}{M} B_{ij} \pi_j,\)</span></p>
<p>where <span class="math notranslate nohighlight">\(M\)</span> is an overall scaling factor chosen so that the largest
row sum in the matrix <span class="math notranslate nohighlight">\(B_{ij} \pi_j /M\)</span> is equal to one.
In this model, <span class="math notranslate nohighlight">\(M = 1.78314862248\)</span>.
This implementation as a <a class="reference internal" href="#msprime.MatrixMutationModel" title="msprime.MatrixMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixMutationModel</span></code></a> has transition matrix
<span class="math notranslate nohighlight">\(P_{ij} = B_{ij} \pi_{j} / M\)</span>, and <span class="math notranslate nohighlight">\(P_{ii}\)</span> chosen so rows sum to one,
and root distribution equal to the equilibrium frequencies, <span class="math notranslate nohighlight">\(\pi\)</span>,
so the matrix <span class="math notranslate nohighlight">\(B\)</span> can be recovered as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">PAM</span><span class="p">()</span>
<span class="n">M</span> <span class="o">=</span> <span class="mf">1.78314862248</span>
<span class="n">B</span> <span class="o">=</span> <span class="n">M</span> <span class="o">*</span> <span class="n">model</span><span class="o">.</span><span class="n">transition_matrix</span> <span class="o">/</span> <span class="n">model</span><span class="o">.</span><span class="n">root_distribution</span>
</pre></div>
</div>
<p>Citation: The values of <span class="math notranslate nohighlight">\(B\)</span> and <span class="math notranslate nohighlight">\(\pi\)</span> used here were copied from
<a class="reference external" href="http://tree.bio.ed.ac.uk/software/seqgen/">Seq-Gen</a>,
and follow the “Dayhoff DCMut” model as described in <em>Kosiol, C., and Goldman,
N. 2005. Different versions of the Dayhoff rate matrix. Molecular Biology and
Evolution 22:193-199.</em> The original citation is: <em>Dayhoff, M.O., Schwartz,
R.M., Orcutt, B.C. (1978). A model of evolutionary change in proteins. Atlas of
Protein Sequence Structures, Vol. 5, Suppl. 3, National Biomedical Research
Foundation, Washington DC, pp. 345-352.</em></p>
</dd></dl>

<dl class="py class">
<dt id="msprime.MatrixMutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">MatrixMutationModel</code><a class="headerlink" href="#msprime.MatrixMutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Superclass of the specific mutation models with a finite set of states.
This class can be used to create a matrix mutation model with custom
alleles and probabilities. For details of how this works see
<a class="reference internal" href="mutations.html#sec-mutations-matrix-mutation-models-details"><span class="std std-ref">Mutation Matrix Models Details</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alleles</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – A list of the possible alleles generated by the model. Each
entry is a string.</p></li>
<li><p><strong>root_distribution</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – An array of float values the same length
as <code class="docutils literal notranslate"><span class="pre">alleles</span></code> which should sum to 1. These values are used to determine the
ancestral state of each mutated site. For example if <code class="docutils literal notranslate"><span class="pre">alleles</span></code> is
<code class="docutils literal notranslate"><span class="pre">['A','B']</span></code> and <code class="docutils literal notranslate"><span class="pre">root_distribution</span></code> is <code class="docutils literal notranslate"><span class="pre">[1,0]</span></code> all ancestral states will be
<code class="docutils literal notranslate"><span class="pre">'A'</span></code>.</p></li>
<li><p><strong>transition_matrix</strong> (<em>array-like</em><em>[</em><em>float</em><em>]</em>) – A square 2d-matrix of transition
probabilities where both dimensions are equal to the length of the <code class="docutils literal notranslate"><span class="pre">alleles</span></code>
argument. These are used to determine the derived state at each mutation. Note
that rows should sum to 1.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.InfiniteAllelesMutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">InfiniteAllelesMutationModel</code><a class="headerlink" href="#msprime.InfiniteAllelesMutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>An <em>infinite alleles</em> model of mutation. This works by keeping track of a “next
allele”: each time the model is asked to produce a new allele (either for an
ancestral or derived state), the “next allele” is provided, and then
incremented.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>start_allele</strong> (<em>int</em>) – The nonnegative integer to start assigning alleles from.
(default: 0)</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><p><strong>next_allele</strong> (<em>int</em>) – The <em>next</em> allele to be assigned. This increments by
one each time an allele is assigned to a new mutation, and resets to
<code class="docutils literal notranslate"><span class="pre">start_allele</span></code> each time a new ancestral state is assigned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.SLiMMutationModel">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">SLiMMutationModel</code><a class="headerlink" href="#msprime.SLiMMutationModel" title="Permalink to this definition">¶</a></dt>
<dd><p>An infinite-alleles model of mutation producing “SLiM-style” mutations.</p>
<p>To agree with mutations produced by SLiM, the ancestral state of each new
site is set to the empty string, and each derived state is produced by
appending the “next allele” to the previous state. The result is that each
allele is a comma-separated string of all mutations that have occurred up
to the root.</p>
<p>Mutations produced by SLiM carry both a <code class="docutils literal notranslate"><span class="pre">time</span></code> attribute, in units of “time ago”
as usual for tskit, as well as a metadata attributed called “origin_generation”,
in units of time since the start of the simulation. Adding these two together -
time since the start of the simulation plus time until the end - is equal to the
total number of generations of the simulation. The origin_generation is not
currently used by SLiM, but for consistency, the origin_genration attribute for
mutations produced by this model is set equal to <code class="docutils literal notranslate"><span class="pre">slim_generation</span></code> minus
<code class="docutils literal notranslate"><span class="pre">floor(mut.time)</span></code>, where <code class="docutils literal notranslate"><span class="pre">mut.time</span></code> is the (tskit) time ago of the mutation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>type</strong> (<em>int</em>) – The nonnegative integer defining the “type” of SLiM mutation
that will be recorded in metadata.</p></li>
<li><p><strong>next_id</strong> (<em>int</em>) – The nonnegative integer to start assigning alleles from.
(default: 0)</p></li>
<li><p><strong>slim_generation</strong> (<em>int</em>) – The “SLiM generation” time used in determining the
“origin_generation” metadata attribute of mutations. This can usually
be left at its default, which is 1.</p></li>
<li><p><strong>block_size</strong> (<em>int</em>) – The block size for allocating derived states.
You do not need to change this unless you get an “out of memory” error
due to a very large number of stacked mutations.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="node-flags">
<span id="sec-api-node-flags"></span><h2>Node flags<a class="headerlink" href="#node-flags" title="Permalink to this headline">¶</a></h2>
<p>For standard coalescent simulations, all samples are marked with the
<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.NODE_IS_SAMPLE" title="(in tskit v0.3)"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NODE_IS_SAMPLE</span></code></a> flag; internal nodes all have a flags value of 0.</p>
<!---
todo link these up with the examples sections below where they are used.
-->
<dl class="py data">
<dt id="msprime.msprime.NODE_IS_RE_EVENT">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">NODE_IS_RE_EVENT</code><a class="headerlink" href="#msprime.msprime.NODE_IS_RE_EVENT" title="Permalink to this definition">¶</a></dt>
<dd><p>The node is an ARG recombination event. Each recombination event is marked
with two nodes, one identifying the individual providing the genetic
material to the left of the breakpoint and the other providing the genetic
material the right. Only present if the <code class="docutils literal notranslate"><span class="pre">record_full_arg</span></code> option is
specified.</p>
</dd></dl>

<dl class="py data">
<dt id="msprime.msprime.NODE_IS_CA_EVENT">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">NODE_IS_CA_EVENT</code><a class="headerlink" href="#msprime.msprime.NODE_IS_CA_EVENT" title="Permalink to this definition">¶</a></dt>
<dd><p>The node is an ARG common ancestor event that did not result in
marginal coalescence. Only present if the <code class="docutils literal notranslate"><span class="pre">record_full_arg</span></code> option is
specified.</p>
</dd></dl>

<dl class="py data">
<dt id="msprime.msprime.NODE_IS_MIG_EVENT">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">NODE_IS_MIG_EVENT</code><a class="headerlink" href="#msprime.msprime.NODE_IS_MIG_EVENT" title="Permalink to this definition">¶</a></dt>
<dd><p>The node is an ARG migration event identifying the individual that migrated.
Can be used in combination with the <code class="docutils literal notranslate"><span class="pre">record_migrations</span></code>.
Only present if the <code class="docutils literal notranslate"><span class="pre">record_full_arg</span></code> option is
specified.</p>
</dd></dl>

<dl class="py data">
<dt id="msprime.msprime.NODE_IS_CEN_EVENT">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">NODE_IS_CEN_EVENT</code><a class="headerlink" href="#msprime.msprime.NODE_IS_CEN_EVENT" title="Permalink to this definition">¶</a></dt>
<dd><p>The node was created by a :class:<code class="docutils literal notranslate"><span class="pre">msprime.CensusEvent</span></code>.</p>
</dd></dl>

</div>
<div class="section" id="utilities">
<h2>Utilities<a class="headerlink" href="#utilities" title="Permalink to this headline">¶</a></h2>
<div class="section" id="rate-maps">
<h3>Rate maps<a class="headerlink" href="#rate-maps" title="Permalink to this headline">¶</a></h3>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Add some high-level content here.</p>
</div>
<dl class="py class">
<dt id="msprime.RateMap">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">RateMap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span></em>, <em class="sig-param"><span class="n">rate</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">start_position</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end_position</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RateMap" title="Permalink to this definition">¶</a></dt>
<dd><p>A class mapping a numeric value to a set of adjacent intervals along the
genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<em>list</em>) – A list of <span class="math notranslate nohighlight">\(n+1\)</span> positions, starting at 0, and ending
in the sequence length over which the RateMap will apply.</p></li>
<li><p><strong>rate</strong> (<em>list</em>) – A list of <span class="math notranslate nohighlight">\(n\)</span> positive rates that apply between each
position.</p></li>
<li><p><strong>start_position</strong> (<em>float</em>) – When returning a mean_rate over the genome, ignore the
part of the map from zero up to <code class="docutils literal notranslate"><span class="pre">start_position</span></code>, as long as the rate is 0 in
that region. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), the <code class="docutils literal notranslate"><span class="pre">start_position</span></code> is 0.</p></li>
<li><p><strong>end_position</strong> (<em>float</em>) – When returning a mean_rate over the genome, ignore the
part of the map from <code class="docutils literal notranslate"><span class="pre">end_position</span></code> onwards, as long as the rate is 0 in that
region. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), the <code class="docutils literal notranslate"><span class="pre">end_position</span></code> is the end of the map.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="msprime.RateMap.cumulative_mass">
<em class="property">property </em><code class="sig-name descname">cumulative_mass</code><a class="headerlink" href="#msprime.RateMap.cumulative_mass" title="Permalink to this definition">¶</a></dt>
<dd><p>The integral of the rates along the genome, one value for each
<a class="reference internal" href="#msprime.RateMap.position" title="msprime.RateMap.position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">position</span></code></a>. The first value will be 0 and the last will be the
<a class="reference internal" href="#msprime.RateMap.total_mass" title="msprime.RateMap.total_mass"><code class="xref py py-attr docutils literal notranslate"><span class="pre">total_mass</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.end_position">
<em class="property">property </em><code class="sig-name descname">end_position</code><a class="headerlink" href="#msprime.RateMap.end_position" title="Permalink to this definition">¶</a></dt>
<dd><p>The physical end of the map. This position can be less than the
<a class="reference internal" href="#msprime.RateMap.sequence_length" title="msprime.RateMap.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> for maps which have no recorded rate at the end of a
chromosome (often the telomeric region). It primarily affects the
<a class="reference internal" href="#msprime.RateMap.mean_rate" title="msprime.RateMap.mean_rate"><code class="xref py py-attr docutils literal notranslate"><span class="pre">mean_rate</span></code></a> returned for this rate map.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.get_cumulative_mass">
<code class="sig-name descname">get_cumulative_mass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">x</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RateMap.get_cumulative_mass" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the cumulative rate at a list of user-specified positions along the map,
i.e. interpolated between the fixed <a class="reference internal" href="#msprime.RateMap.position" title="msprime.RateMap.position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">position</span></code></a> values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>x</strong> (<em>numpy.ndarray</em>) – The positions for which to return values.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An array of cumulative rates, the same length as <code class="docutils literal notranslate"><span class="pre">x</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.ndarray(dtype=np.float64)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.get_rate">
<code class="sig-name descname">get_rate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">x</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RateMap.get_rate" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the rate at a list of user-specified positions along the map, i.e.
interpolated between the fixed <a class="reference internal" href="#msprime.RateMap.position" title="msprime.RateMap.position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">position</span></code></a> values. Any positions
beyond the end  of the map will return the final rate value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>x</strong> (<em>numpy.ndarray</em>) – The positions for which to return values.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An array of rates, the same length as <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.ndarray(dtype=np.float64)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.mean_rate">
<em class="property">property </em><code class="sig-name descname">mean_rate</code><a class="headerlink" href="#msprime.RateMap.mean_rate" title="Permalink to this definition">¶</a></dt>
<dd><p>The mean rate over this map, from <a class="reference internal" href="#msprime.RateMap.start_position" title="msprime.RateMap.start_position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">start_position</span></code></a> to
<a class="reference internal" href="#msprime.RateMap.end_position" title="msprime.RateMap.end_position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">end_position</span></code></a>, weighted by the span covered by each rate.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.position">
<em class="property">property </em><code class="sig-name descname">position</code><a class="headerlink" href="#msprime.RateMap.position" title="Permalink to this definition">¶</a></dt>
<dd><p>The positions between which a constant rate applies. There will be
one more position than the number of rates, with the first position being 0
and the last being the <a class="reference internal" href="#msprime.RateMap.sequence_length" title="msprime.RateMap.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.rate">
<em class="property">property </em><code class="sig-name descname">rate</code><a class="headerlink" href="#msprime.RateMap.rate" title="Permalink to this definition">¶</a></dt>
<dd><p>The array of constant rates that apply between each <a class="reference internal" href="#msprime.RateMap.position" title="msprime.RateMap.position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">position</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.read_hapmap">
<em class="property">static </em><code class="sig-name descname">read_hapmap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fileobj</span></em>, <em class="sig-param"><span class="n">sequence_length</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">has_header</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">position_col</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rate_col</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">map_col</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RateMap.read_hapmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the specified file in HapMap format and returns a RateMap.
HapMap files must white-space-delimited, and by default are assumed to
contain a single header line (which is ignored). Each subsequent line
then contains a physical position (in base pairs) and either a genetic
map position (in centiMorgans) or a recombination rate (in centiMorgans
per megabase); the rate between the current physical position
(inclusive) and the physical position on the next line (exclusive) is
taken as constant. By default, the second column of the file is taken
as the physical position and the fourth column is taken as the genetic
position, as seen in the following sample of the format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Chromosome</span>  <span class="n">Position</span><span class="p">(</span><span class="n">bp</span><span class="p">)</span>  <span class="n">Rate</span><span class="p">(</span><span class="n">cM</span><span class="o">/</span><span class="n">Mb</span><span class="p">)</span>  <span class="n">Map</span><span class="p">(</span><span class="n">cM</span><span class="p">)</span>
<span class="n">chr10</span>       <span class="mi">48232</span>         <span class="mf">0.1614</span>       <span class="mf">0.002664</span>
<span class="n">chr10</span>       <span class="mi">48486</span>         <span class="mf">0.1589</span>       <span class="mf">0.002705</span>
<span class="n">chr10</span>       <span class="mi">50009</span>         <span class="mf">0.159</span>        <span class="mf">0.002947</span>
<span class="n">chr10</span>       <span class="mi">52147</span>         <span class="mf">0.1574</span>       <span class="mf">0.003287</span>
<span class="o">...</span>
<span class="n">chr10</span>       <span class="mi">133762002</span>     <span class="mf">3.358</span>        <span class="mf">181.129345</span>
<span class="n">chr10</span>       <span class="mi">133766368</span>     <span class="mf">0.000</span>        <span class="mf">181.144008</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The rows are all assumed to come from the same contig, and the
first column is currently ignored. Therefore if you have a single
file containing several contigs or chromosomes, you must must split
it up into multiple files, and pass each one separately to this
function.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fileobj</strong> (<em>str</em>) – Filename or file to read. This is passed directly
to <code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.loadtxt()</span></code>, so if the filename extension is .gz or .bz2,
the file is decompressed first</p></li>
<li><p><strong>sequence_length</strong> (<em>float</em>) – The total length of the map. If <code class="docutils literal notranslate"><span class="pre">None</span></code>,
then assume it is the last physical position listed in the file.
Otherwise it must be greater then or equal to the last physical
position in the file, and the region between the last physical position
and the sequence_length is padded with a rate of zero.</p></li>
<li><p><strong>has_header</strong> (<em>bool</em>) – If True (default), assume the file has a header row
and ignore the first line of the file.</p></li>
<li><p><strong>position_col</strong> (<em>int</em>) – The zero-based index of the column in the file
specifying the physical position in base pairs. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default)
assume an index of 1 (i.e. the second column).</p></li>
<li><p><strong>rate_col</strong> (<em>int</em>) – The zero-based index of the column in the file
specifying the rate in cM/Mb. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) do not use the rate
column, but calculate rates using the genetic map positions, as
specified in <code class="docutils literal notranslate"><span class="pre">map_col</span></code>. If the rate column is used, the
<a class="reference internal" href="#msprime.RateMap.start_position" title="msprime.RateMap.start_position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">RateMap.start_position</span></code></a> of the returned map is set to the first
physical position in the file, and the last value in the rate column
must be zero.</p></li>
<li><p><strong>map_col</strong> (<em>int</em>) – The zero-based index of the column in the file
specifying the genetic map position in centiMorgans. If <code class="docutils literal notranslate"><span class="pre">None</span></code>
(default), assume an index of 3 (i.e. the fourth column). If the first
genetic position is 0, set the <a class="reference internal" href="#msprime.RateMap.start_position" title="msprime.RateMap.start_position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">RateMap.start_position</span></code></a> of the
returned map to the first physical position in the file. Otherwise, act
as if an additional row, specifying physical position 0 and genetic
position 0, exists at the start of the file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A RateMap object.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.RateMap" title="msprime.RateMap">RateMap</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.sequence_length">
<em class="property">property </em><code class="sig-name descname">sequence_length</code><a class="headerlink" href="#msprime.RateMap.sequence_length" title="Permalink to this definition">¶</a></dt>
<dd><p>The sequence length covered by this map</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.slice">
<code class="sig-name descname">slice</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">trim</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RateMap.slice" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a subset of this rate map between the specified end points.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>float</em>) – The start of the region to keep. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, it defaults to
0.</p></li>
<li><p><strong>end</strong> (<em>float</em>) – The end of the region to keep. If end is <code class="docutils literal notranslate"><span class="pre">None</span></code>, it defaults
to the end of the map.</p></li>
<li><p><strong>trim</strong> (<em>bool</em>) – If True, remove the flanking regions such that the
sequence length of the new rate map is <code class="docutils literal notranslate"><span class="pre">end</span></code> - <code class="docutils literal notranslate"><span class="pre">start</span></code>. If <code class="docutils literal notranslate"><span class="pre">False</span></code>
(default), do not change the coordinate system, but instead replace each
flanking region with 0, and set the <code class="docutils literal notranslate"><span class="pre">start_position</span></code> and <code class="docutils literal notranslate"><span class="pre">end_position</span></code>
of the returned map to <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code>, inserting these positions if
they do not exist.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new RateMap instance</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.RateMap" title="msprime.RateMap">RateMap</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.start_position">
<em class="property">property </em><code class="sig-name descname">start_position</code><a class="headerlink" href="#msprime.RateMap.start_position" title="Permalink to this definition">¶</a></dt>
<dd><p>The physical start of the map. This position can be greater than 0 for maps
which have no recorded rate at the start of a chromosome (often the telomeric
region). It primarily affects the <a class="reference internal" href="#msprime.RateMap.mean_rate" title="msprime.RateMap.mean_rate"><code class="xref py py-attr docutils literal notranslate"><span class="pre">mean_rate</span></code></a> returned for this rate map.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.total_mass">
<em class="property">property </em><code class="sig-name descname">total_mass</code><a class="headerlink" href="#msprime.RateMap.total_mass" title="Permalink to this definition">¶</a></dt>
<dd><p>The cumulative total over the entire map</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RateMap.uniform">
<em class="property">static </em><code class="sig-name descname">uniform</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence_length</span></em>, <em class="sig-param"><span class="n">rate</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RateMap.uniform" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a uniform rate map</p>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="id2">
<h2>Demography<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="msprime.Demography">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">Demography</code><span class="sig-paren">(</span><em class="sig-param">populations: List[msprime.demography.Population] = &lt;factory&gt;</em>, <em class="sig-param">events: List = &lt;factory&gt;</em>, <em class="sig-param">migration_matrix: Optional[numpy.ndarray] = None</em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography" title="Permalink to this definition">¶</a></dt>
<dd><p>A description of a demographic model for an msprime simulation.</p>
<p>TODO document properly.</p>
<dl class="py method">
<dt id="msprime.Demography.debug">
<code class="sig-name descname">debug</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.debug" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> instance for this demography.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A DemographyDebugger object for this demography.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#msprime.DemographyDebugger" title="msprime.DemographyDebugger">DemographyDebugger</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.from_old_style">
<em class="property">static </em><code class="sig-name descname">from_old_style</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">population_configurations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">migration_matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">demographic_events</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">Ne</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.from_old_style" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a Demography object from the pre 1.0 style input parameters,
reproducing the old semantics with respect to default values.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.from_species_tree">
<em class="property">static </em><code class="sig-name descname">from_species_tree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tree</span></em>, <em class="sig-param"><span class="n">initial_size</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">branch_length_units</span><span class="o">=</span><span class="default_value">'gen'</span></em>, <em class="sig-param"><span class="n">generation_time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.from_species_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a species tree in <a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">Newick</a> format and return the
corresponding <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object. The tree is assumed to be
rooted and ultrametric and branch lengths must be included and
correspond to time, either in units of millions of years, years, or
generations.</p>
<p>The returned <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object contains a
<a class="reference internal" href="#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> for each node in the species tree. The
population’s <code class="docutils literal notranslate"><span class="pre">name</span></code> attribute will be either the corresponding node
label from the newick tree, if it exists, or otherwise the name takes
the form “pop_{j}”, where j is the position of the given population in
the list. Leaf populations are first in the list, and added in
left-to-right order. Populations corresponding to the internal nodes
are then added in a postorder traversal of the species tree. For each
internal node, <a class="reference internal" href="#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-class docutils literal notranslate"><span class="pre">MassMigration</span></code></a> events are added so that
lineages move from its child populations at the appropriate time.</p>
<dl class="field-list simple">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>If continuous migration is added to the returned Demography
by updating the migration matrix, it is important to
note that the MassMigration events used to move lineages do <em>not</em>
alter migration rates, and it should be ensured that
migration rates to source populations of mass migration events are zero
after the mass migration (viewed backwards in time).</p>
</dd>
<dt class="field-even">Todo</dt>
<dd class="field-even"><p>Implement the PopulationSplit event and document its use here.
We’ll still need to put in a warning, so users know that the migration
matrix will be updated at every node.</p>
</dd>
</dl>
<p>The initial sizes and growth rates for the populations in the model are
set via the <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> and <code class="docutils literal notranslate"><span class="pre">growth_rate</span></code> arguments. These can be
specified in two ways: if a single number is provided, this is used
for all populations. The argument may also be a mapping from population
names to their respective values. For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;(A:10.0,B:10.0)C&quot;</span>
<span class="n">initial_size</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">2000</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">}</span>
<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_species_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">initial_size</span><span class="p">)</span>
</pre></div>
</div>
<p>Note that it is possible to have default population sizes for unnamed
ancestral populations using a <cite>collections.defaultdict</cite>, e.g.,</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;(A:10.0,B:10.0)&quot;</span>
<span class="n">initial_size</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">initial_size</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">2000</span><span class="p">})</span>
<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_species_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">initial_size</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<em>str</em>) – The tree string in Newick format, with named leaves and branch
lengths.</p></li>
<li><p><strong>initial_size</strong> – Each population’s initial_size. May be a single number
or a mapping from population names to their sizes.</p></li>
<li><p><strong>growth_rate</strong> – Each population’s growth_rate. May be a single number
or a mapping from population names to their exponential growth rates.
Defaults to zero.</p></li>
<li><p><strong>branch_length_units</strong> (<em>str</em>) – The units of time in which the species tree’s
branch lengths are measured. Allowed branch length units are millions of
years, years, and generations; these should be specified with the strings
<code class="docutils literal notranslate"><span class="pre">&quot;myr&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;yr&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;gen&quot;</span></code>, respectively. This defaults to
<code class="docutils literal notranslate"><span class="pre">&quot;gen&quot;</span></code>.</p></li>
<li><p><strong>generation_time</strong> (<em>float</em>) – The number of years per generation. If and only
if the branch lengths are not in units of generations, the generation time
must be specified. This defaults to <cite>None</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Demography object representing the specified species tree.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.Demography" title="msprime.Demography">Demography</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.from_starbeast">
<em class="property">static </em><code class="sig-name descname">from_starbeast</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tree</span></em>, <em class="sig-param"><span class="n">generation_time</span></em>, <em class="sig-param"><span class="n">branch_length_units</span><span class="o">=</span><span class="default_value">'myr'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.from_starbeast" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a species tree produced by the program <a class="reference external" href="https://www.beast2.org/treeannotator">TreeAnnotator</a>
based on a posterior tree distribution generated with <a class="reference external" href="https://academic.oup.com/mbe/article/34/8/2101/3738283">StarBEAST</a>  and return
the corresponding Demography object.</p>
<p>Species trees produced by TreeAnnotator are written in <a class="reference external" href="https://en.wikipedia.org/wiki/Nexus_file">Nexus</a> format and are rooted,
bifurcating, and ultrametric. Branch lengths usually are in units of
millions of years, but the use of other units is permitted by StarBEAST
(and thus TreeAnnotator). This function allows branch length units of
millions of years or years. Leaves must be named and the tree must
include information on population sizes of leaf and ancestral species
in the form of annotation with the “dmv” tag, which is the case for
trees written by TreeAnnotator based on StarBEAST posterior tree
distributions.</p>
<p>The returned <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object contains a <a class="reference internal" href="#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> for
each node in the species tree. The population’s <code class="docutils literal notranslate"><span class="pre">name</span></code> attribute will
be either the corresponding node label from the newick tree, if it exists,
or otherwise the name takes the form “pop_{j}”, where j is the position
of the given population in the list. Leaf populations are first in the
list, and added in left-to-right order. Populations corresponding to the
internal nodes are then added in a postorder traversal of the species
tree. For each internal node, <a class="reference internal" href="#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-class docutils literal notranslate"><span class="pre">MassMigration</span></code></a> events are added
so that lineages move from its child populations at the appropriate time.</p>
<dl class="field-list simple">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>If continuous migration is added to the returned Demography
by updating the migration matrix, it is important to
note that the MassMigration events used to move lineages do <em>not</em>
alter migration rates, and it should be ensured that
migration rates to source populations of mass migration events are zero
after the mass migration (viewed backwards in time).</p>
</dd>
<dt class="field-even">Todo</dt>
<dd class="field-even"><p>Implement the PopulationSplit event and document its use here.
We’ll still need to put in a warning, so users know that the migration
matrix will be updated at every node.</p>
</dd>
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<em>str</em>) – The tree string in Nexus format, with named leaves, branch
lengths, and branch annotation. Typically, this string is the entire content
of a file written by TreeAnnotator.</p></li>
<li><p><strong>generation_time</strong> (<em>float</em>) – The number of years per generation.</p></li>
<li><p><strong>branch_length_units</strong> (<em>str</em>) – The units of time in which the species tree’s
branch lengths are measured. Allowed branch length units are millions of
years, and years; these should be specified with the strings <code class="docutils literal notranslate"><span class="pre">&quot;myr&quot;</span></code> or
<code class="docutils literal notranslate"><span class="pre">&quot;yr&quot;</span></code>, respectively. This defaults to <code class="docutils literal notranslate"><span class="pre">&quot;myr&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> instance that describing the information in the
specified species tree.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.Demography" title="msprime.Demography">Demography</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.insert_populations">
<code class="sig-name descname">insert_populations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tables</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.insert_populations" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert population definitions for this demography into the specified
set of tables.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.island_model">
<em class="property">static </em><code class="sig-name descname">island_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">initial_size</span></em>, <em class="sig-param"><span class="n">migration_rate</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.island_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object representing a collection of
populations with specified initial population sizes and growth
rates, with symmetric migration between each pair of populations at the
specified rate. Please see <a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">Demography</span></a> for more details on
population sizes and growth rates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>initial_size</strong> (<em>array_like</em>) – the <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> value for each
of the <a class="reference internal" href="#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> in the returned model. The length
of the array corresponds to the number of populations.
model.</p></li>
<li><p><strong>migration_rate</strong> (<em>float</em>) – The migration rate between each pair of
populations.</p></li>
<li><p><strong>growth_rate</strong> (<em>array_like</em>) – The exponential growth rate for each
population. Must be either None (the default, resulting a zero
growth rate) or an array with the same length as <code class="docutils literal notranslate"><span class="pre">initial_size</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Demography object representing this model, suitable as
input to <a class="reference internal" href="legacy.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.Demography" title="msprime.Demography">Demography</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.isolated_model">
<em class="property">static </em><code class="sig-name descname">isolated_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">initial_size</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.isolated_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object representing a collection of
isolated populations with specified initial population sizes and
growth rates. Please see <a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">Demography</span></a> for more details on
population sizes and growth rates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>initial_size</strong> (<em>array_like</em>) – the <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> value for each
of the <a class="reference internal" href="#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> in the returned model. The length
of the array corresponds to the number of populations.
model.</p></li>
<li><p><strong>growth_rate</strong> (<em>array_like</em>) – The exponential growth rate for each
population. Must be either None (the default, resulting a zero
growth rate) or an array with the same length as <code class="docutils literal notranslate"><span class="pre">initial_size</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Demography object representing this model, suitable as
input to <a class="reference internal" href="legacy.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.Demography" title="msprime.Demography">Demography</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.name_to_id">
<code class="sig-name descname">name_to_id</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.name_to_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the integer ID (i.e., its position in the list of populations)
of the population with the specified name. If the name does not exist,
raise a KeyError.</p>
<p>Note: this function will raise an error if called before the <code class="docutils literal notranslate"><span class="pre">validate</span></code>
method is called.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – The name of the population we wish to look up.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The integer ID of the population.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.stepping_stone_model">
<em class="property">static </em><code class="sig-name descname">stepping_stone_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">initial_size</span></em>, <em class="sig-param"><span class="n">migration_rate</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">boundaries</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.stepping_stone_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object representing a collection of
populations with specified initial population sizes and growth
rates, in which adjacent demes exchange migrants at the
specified rate. Please see <a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">Demography</span></a> for more details on
population sizes and growth rates.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The current implementation on supports a one-dimensional stepping
stone model, but higher dimensions could also be supported. Please
open an issue on GitHub if this feature would be useful to you.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>initial_size</strong> (<em>array_like</em>) – the <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> value for each
of the <a class="reference internal" href="#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> in the returned model. The length
of the array corresponds to the number of populations.</p></li>
<li><p><strong>migration_rate</strong> (<em>float</em>) – The migration rate between adjacent pairs
of populations.</p></li>
<li><p><strong>growth_rate</strong> (<em>array_like</em>) – The exponential growth rate for each
population. Must be either None (the default, resulting a zero
growth rate) or an array with the same length as <code class="docutils literal notranslate"><span class="pre">initial_size</span></code>.</p></li>
<li><p><strong>boundaries</strong> (<em>bool</em>) – If True the stepping stone model has boundary
conditions imposed so that demes at either end of the chain do
not exchange migrats. If False (the default), the set of
populations is “circular” and migration takes place between the
terminal demes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Demography object representing this model, suitable as
input to <a class="reference internal" href="legacy.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.Demography" title="msprime.Demography">Demography</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.validate">
<code class="sig-name descname">validate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Demography.validate" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the demography looks sensible and raises errors/warnings
appropriately.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="msprime.Population">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">Population</code><span class="sig-paren">(</span><em class="sig-param">initial_size: float</em>, <em class="sig-param">growth_rate: float = 0.0</em>, <em class="sig-param">name: Optional[str] = None</em>, <em class="sig-param">description: Optional[str] = None</em>, <em class="sig-param">extra_metadata: dict = &lt;factory&gt;</em>, <em class="sig-param">sampling_time: float = 0</em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Population" title="Permalink to this definition">¶</a></dt>
<dd><p>Define a single population in a simulation.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>initial_size</strong> (<em>float</em>) – The absolute size of the population at time zero.</p></li>
<li><p><strong>growth_rate</strong> (<em>float</em>) – The exponential growth rate of the
population per generation (forwards in time).
Growth rates can be negative. This is zero for a
constant population size, and positive for a population that has been
growing. Defaults to 0.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – The name of the population. If specified this must be a uniquely
identifying string.</p></li>
<li><p><strong>description</strong> (<em>str</em>) – a short description of the population</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="msprime.Population.from_old_style">
<em class="property">static </em><code class="sig-name descname">from_old_style</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pop_config</span></em>, <em class="sig-param"><span class="n">Ne</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.Population.from_old_style" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Population object derived from the specified old-style
PopulationConfiguration. The Ne value is used as the <code class="docutils literal notranslate"><span class="pre">initial_size</span></code>
if this is not provided in the PopulationConfiguration.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="msprime.PopulationParametersChange">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">PopulationParametersChange</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">initial_size</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">population</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>None<span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">population_id</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>None<span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.PopulationParametersChange" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes the demographic parameters of a population at a given time.</p>
<p>This event generalises the <code class="docutils literal notranslate"><span class="pre">-eg</span></code>, <code class="docutils literal notranslate"><span class="pre">-eG</span></code>, <code class="docutils literal notranslate"><span class="pre">-en</span></code> and <code class="docutils literal notranslate"><span class="pre">-eN</span></code>
options from <code class="docutils literal notranslate"><span class="pre">ms</span></code>. Note that unlike <code class="docutils literal notranslate"><span class="pre">ms</span></code> we do not automatically
set growth rates to zero when the population size is changed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<em>float</em>) – The length of time ago at which this event
occurred.</p></li>
<li><p><strong>initial_size</strong> (<em>float</em>) – The absolute diploid size of the population
at the beginning of the time slice starting at <code class="docutils literal notranslate"><span class="pre">time</span></code>. If None,
this is calculated according to the initial population size and
growth rate over the preceding time slice.</p></li>
<li><p><strong>growth_rate</strong> (<em>float</em>) – The new per-generation growth rate. If None,
the growth rate is not changed. Defaults to None.</p></li>
<li><p><strong>population</strong> (<em>int</em>) – The ID of the population affected. If
<code class="docutils literal notranslate"><span class="pre">population</span></code> is None, the changes affect all populations
simultaneously.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.MigrationRateChange">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">MigrationRateChange</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">rate</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">source</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">dest</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">matrix_index</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>tuple<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.MigrationRateChange" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes the rate of migration from one deme to another to a new value at a
specific time. Migration rates are specified in terms of the rate at which
lineages move from population <code class="docutils literal notranslate"><span class="pre">source</span></code> to <code class="docutils literal notranslate"><span class="pre">dest</span></code> during the progress of
the simulation. Note that <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code> are from the perspective
of the coalescent process; please see the <a class="reference internal" href="ancestry.html#sec-ancestry-models"><span class="std std-ref">Models</span></a>
section for more details on the interpretation of this migration model.</p>
<p>By default, <code class="docutils literal notranslate"><span class="pre">source=-1</span></code> and <code class="docutils literal notranslate"><span class="pre">dest=-1</span></code>, which results in all
non-diagonal elements of the migration matrix being changed to the new
rate. If <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code> are specified, they must refer to valid
population IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<em>float</em>) – The time at which this event occurs in generations.</p></li>
<li><p><strong>rate</strong> (<em>float</em>) – The new per-generation migration rate.</p></li>
<li><p><strong>source</strong> (<em>int</em>) – The ID of the source population.</p></li>
<li><p><strong>dest</strong> (<em>int</em>) – The ID of the destination population.</p></li>
<li><p><strong>source</strong> – The source population ID.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.MassMigration">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">MassMigration</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">source</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">dest</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>None<span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">proportion</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">destination</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>None<span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.MassMigration" title="Permalink to this definition">¶</a></dt>
<dd><p>A mass migration event in which some fraction of the population in one deme
(the <code class="docutils literal notranslate"><span class="pre">source</span></code>) simultaneously move to another deme (<code class="docutils literal notranslate"><span class="pre">dest</span></code>) during the
progress of the simulation. Each lineage currently present in the source
population moves to the destination population with probability equal to
<code class="docutils literal notranslate"><span class="pre">proportion</span></code>. Note that <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code> are from the perspective
of the coalescent process; please see the <a class="reference internal" href="ancestry.html#sec-ancestry-models"><span class="std std-ref">Models</span></a>
section for more details on the interpretation of this migration model.</p>
<p>This event class generalises the population split (<code class="docutils literal notranslate"><span class="pre">-ej</span></code>) and
admixture (<code class="docutils literal notranslate"><span class="pre">-es</span></code>) events from <code class="docutils literal notranslate"><span class="pre">ms</span></code>. Note that MassMigrations
do <em>not</em> have any side effects on the migration matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<em>float</em>) – The time at which this event occurs in generations.</p></li>
<li><p><strong>source</strong> (<em>int</em>) – The ID of the source population.</p></li>
<li><p><strong>dest</strong> (<em>int</em>) – The ID of the destination population.</p></li>
<li><p><strong>proportion</strong> (<em>float</em>) – The probability that any given lineage within
the source population migrates to the destination population.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.CensusEvent">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">CensusEvent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span><span class="p">:</span> <span class="n">float</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.CensusEvent" title="Permalink to this definition">¶</a></dt>
<dd><p>An event that adds a node to each branch of every tree at a given time
during the simulation. This may be used to record all ancestral haplotypes
present at that time, and to extract other information related to these
haplotypes: for instance to trace the local ancestry of a sample back to a
set of contemporaneous ancestors, or to assess whether a subset of samples
has coalesced more recently than the census time.</p>
<p>See <a class="reference internal" href="ancestry.html#sec-ancestry-census-events"><span class="std std-ref">Census events</span></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>time</strong> (<em>float</em>) – The time at which this event occurs in generations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.DemographyDebugger">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">DemographyDebugger</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">Ne</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">population_configurations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">migration_matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">demographic_events</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">demography</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.DemographyDebugger" title="Permalink to this definition">¶</a></dt>
<dd><p>A class to facilitate debugging of population parameters and migration
rates in the past.</p>
<dl class="py method">
<dt id="msprime.DemographyDebugger.coalescence_rate_trajectory">
<code class="sig-name descname">coalescence_rate_trajectory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">steps</span></em>, <em class="sig-param"><span class="n">num_samples</span></em>, <em class="sig-param"><span class="n">min_pop_size</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">double_step_validation</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.DemographyDebugger.coalescence_rate_trajectory" title="Permalink to this definition">¶</a></dt>
<dd><p>This function will calculate the mean coalescence rates and proportions
of uncoalesced lineages between the lineages of the sample
configuration provided in <cite>num_samples</cite>, at each of the times ago
listed by steps, in this demographic model. The coalescence rate at
time t in the past is the average rate of coalescence of
as-yet-uncoalesed lineages, computed as follows: let <span class="math notranslate nohighlight">\(p(t)\)</span> be
the probability that the lineages of a randomly chosen pair of samples
has not yet coalesced by time <span class="math notranslate nohighlight">\(t\)</span>, let <span class="math notranslate nohighlight">\(p(z,t)\)</span> be the
probability that the lineages of a randomly chosen pair of samples has
not yet coalesced by time <span class="math notranslate nohighlight">\(t\)</span> <em>and</em> are both in population
<span class="math notranslate nohighlight">\(z\)</span>, and let <span class="math notranslate nohighlight">\(N(z,t)\)</span> be the diploid effective population
size of population <span class="math notranslate nohighlight">\(z\)</span> at time <span class="math notranslate nohighlight">\(t\)</span>. Then the mean
coalescence rate at time <span class="math notranslate nohighlight">\(t\)</span> is <span class="math notranslate nohighlight">\(r(t) = (\sum_z p(z,t) /
(2 * N(z,t)) / p(t)\)</span>.</p>
<p>The computation is done by approximating population size trajectories
with piecewise constant trajectories between each of the steps. For
this to be accurate, the distance between the steps must be small
enough so that (a) short epochs (e.g., bottlenecks) are not missed, and
(b) populations do not change in size too much over that time, if they
are growing or shrinking. This function optionally provides a simple
check of this approximation by recomputing the coalescence rates on a
grid of steps twice as fine and throwing a warning if the resulting
values do not match to a relative tolerance of 0.001.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>steps</strong> (<em>list</em>) – The times ago at which coalescence rates will be computed.</p></li>
<li><p><strong>num_samples</strong> (<em>list</em>) – A list of the same length as the number
of populations, so that <cite>num_samples[j]</cite> is the number of sampled
chromosomes in subpopulation <cite>j</cite>.</p></li>
<li><p><strong>min_pop_size</strong> (<em>int</em>) – The smallest allowed population size during
computation of coalescent rates (i.e., coalescence rates are actually
1 / (2 * max(min_pop_size, N(z,t))). Spurious very small population sizes
can occur in models where populations grow exponentially but are unused
before some time in the past, and lead to floating point error.
This should be set to a value smaller than the smallest
desired population size in the model.</p></li>
<li><p><strong>double_step_validation</strong> (<em>bool</em>) – Whether to perform the check that
step sizes are sufficiently small, as described above. This is highly
recommended, and will take at most four times the computation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of arrays whose jth elements, respectively, are the
coalescence rate at the jth time point (denoted r(t[j]) above),
and the probablility that a randomly chosen pair of lineages has
not yet coalesced (denoted p(t[j]) above).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(numpy.array, numpy.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.epoch_times">
<em class="property">property </em><code class="sig-name descname">epoch_times</code><a class="headerlink" href="#msprime.DemographyDebugger.epoch_times" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns array of epoch times defined by the demographic model</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.lineage_probabilities">
<code class="sig-name descname">lineage_probabilities</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">steps</span></em>, <em class="sig-param"><span class="n">sample_time</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.DemographyDebugger.lineage_probabilities" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array such that P[j, a, b] is the probability that a lineage that
started in population a at time sample_time is in population b at time steps[j]
ago.</p>
<p>This function reports sampling probabilities _before_ mass migration events
at a step time, if a mass migration event occurs at one of those times.
Migrations will then effect the next time step.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>steps</strong> (<em>list</em>) – A list of times to compute probabilities.</p></li>
<li><p><strong>sample_time</strong> – The time of sampling of the lineage. For any times in steps
that are more recent than sample_time, the probability of finding the
lineage in any population is zero.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An array of dimension len(steps) by num pops by num_pops.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.mean_coalescence_time">
<code class="sig-name descname">mean_coalescence_time</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_samples</span></em>, <em class="sig-param"><span class="n">min_pop_size</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">steps</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rtol</span><span class="o">=</span><span class="default_value">0.005</span></em>, <em class="sig-param"><span class="n">max_iter</span><span class="o">=</span><span class="default_value">12</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.DemographyDebugger.mean_coalescence_time" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the mean time until coalescence between lineages of two samples drawn
from the sample configuration specified in <cite>num_samples</cite>. This is done using
<a class="reference internal" href="#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">coalescence_rate_trajectory</span></code></a>
to compute the probability that the lineages have not yet coalesced by time <cite>t</cite>,
and using these to approximate <span class="math notranslate nohighlight">\(E[T] = \int_t^\infty P(T &gt; t) dt\)</span>,
where <span class="math notranslate nohighlight">\(T\)</span> is the coalescence time. See
<a class="reference internal" href="#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">coalescence_rate_trajectory</span></code></a>
for more details.</p>
<p>To compute this, an adequate time discretization must be arrived at
by iteratively extending or refining the current discretization.
Debugging information about numerical convergence of this procedure is
logged using the Python <code class="xref py py-mod docutils literal notranslate"><span class="pre">logging</span></code> infrastructure. To make it appear, using
the <code class="xref py py-mod docutils literal notranslate"><span class="pre">daiquiri</span></code> module, do for instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">daiquiri</span>

<span class="n">daiquiri</span><span class="o">.</span><span class="n">setup</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">)</span>
<span class="n">debugger</span><span class="o">.</span><span class="n">mean_coalescence_time</span><span class="p">([</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
<p>will print this debugging information to stderr. Briefly, this outputs
iteration number, mean coalescence time, maximum difference in probabilty
of not having coalesced yet, difference to last coalescence time,
probability of not having coalesced by the final time point, and
whether the last iteration was an extension or refinement.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_samples</strong> (<em>list</em>) – A list of the same length as the number
of populations, so that <cite>num_samples[j]</cite> is the number of sampled
chromosomes in subpopulation <cite>j</cite>.</p></li>
<li><p><strong>min_pop_size</strong> (<em>int</em>) – See <a class="reference internal" href="#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">coalescence_rate_trajectory</span></code></a>.</p></li>
<li><p><strong>steps</strong> (<em>list</em>) – The time discretization to start out with (by default,
picks something based on epoch times).</p></li>
<li><p><strong>rtol</strong> (<em>float</em>) – The relative tolerance to determine mean coalescence time
to (used to decide when to stop subdividing the steps).</p></li>
<li><p><strong>max_iter</strong> (<em>int</em>) – The maximum number of times to subdivide the steps.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The mean coalescence time (a number).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.num_epochs">
<em class="property">property </em><code class="sig-name descname">num_epochs</code><a class="headerlink" href="#msprime.DemographyDebugger.num_epochs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of epochs defined by the demographic model.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.population_size_history">
<em class="property">property </em><code class="sig-name descname">population_size_history</code><a class="headerlink" href="#msprime.DemographyDebugger.population_size_history" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a (num_pops, num_epochs) numpy array giving the starting population size
for each population in each epoch.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.population_size_trajectory">
<code class="sig-name descname">population_size_trajectory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">steps</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.DemographyDebugger.population_size_trajectory" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns an array of per-population effective population sizes,
as defined by the demographic model. These are the <cite>initial_size</cite>
parameters of the model, modified by any population growth rates.
The sizes are computed at the time points given by <cite>steps</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>steps</strong> (<em>list</em>) – List of times ago at which the population
size will be computed.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns a numpy array of population sizes, with one column per
population, whose [i,j]th entry is the size of population
j at time steps[i] ago.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.possible_lineage_locations">
<code class="sig-name descname">possible_lineage_locations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.DemographyDebugger.possible_lineage_locations" title="Permalink to this definition">¶</a></dt>
<dd><p>Given the sampling configuration, this function determines when lineages are
possibly found within each population over epochs defined by demographic events
and sampling times. If no sampling configuration is given, we assume we sample
lineages from every population at time zero. The samples are specified by a list
of msprime Sample objects, so that possible ancient samples may be accounted for.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>samples</strong> (<em>list</em>) – A list of msprime Sample objects, which specify their
populations and times.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns a dictionary with epoch intervals as keys whose values are a
list with length equal to the number of populations with True and False
indicating which populations could possibly contain lineages over that
epoch. The epoch intervals are given by tuples: (epoch start, epoch end).
The first epoch necessarily starts at time 0, and the final epoch has end
time of infinity.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.print_history">
<code class="sig-name descname">print_history</code><span class="sig-paren">(</span><em class="sig-param">output=&lt;_io.TextIOWrapper name='&lt;stdout&gt;' mode='w' encoding='utf-8'&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.DemographyDebugger.print_history" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints a summary of the history of the populations.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="likelihoods">
<h2>Likelihoods<a class="headerlink" href="#likelihoods" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="msprime.log_arg_likelihood">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">log_arg_likelihood</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ts</span></em>, <em class="sig-param"><span class="n">recombination_rate</span></em>, <em class="sig-param"><span class="n">Ne</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.log_arg_likelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the log probability of the stored tree sequence under the Hudson ARG.
An exact expression for this probability is given in equation (1) of
<a class="reference external" href="https://www.genetics.org/content/156/3/1393">Kuhner et al. (2000)</a>.</p>
<p>We assume branch lengths stored in generations, resulting in a coalescence
rate of <span class="math notranslate nohighlight">\(1 / (2 N_e)\)</span> per pair of lineages.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The stored tree sequence must store the full realisation of the ARG,
including all recombination events and all common ancestor events,
regardless of whether the recombinations cause a change in the ancestral
tree or whether the common ancestor events cause coalescence of ancestral
material. See <a class="reference internal" href="ancestry.html#sec-ancestry-full-arg"><span class="std std-ref">Ancestral recombination graph</span></a> for details of this
data structure, and how to generate them using <code class="docutils literal notranslate"><span class="pre">msprime</span></code>.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method only supports continuous genomes.
See <a class="reference internal" href="ancestry.html#sec-ancestry-discrete-genome"><span class="std std-ref">Discrete or continuous?</span></a> for how these can be specified
when simulating tree sequences using <code class="docutils literal notranslate"><span class="pre">msprime</span></code>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ts</strong> (<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><em>tskit.TreeSequence</em></a>) – The tree sequence object.</p></li>
<li><p><strong>recombination_rate</strong> (<em>float</em>) – The per-link, per-generation recombination
probability. Must be non-negative.</p></li>
<li><p><strong>Ne</strong> (<em>float</em>) – The diploid effective population size.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The log probability of the tree sequence under the Hudson ancestral
recombination graph model. If the recombination rate is zero and the tree
sequence contains at least one recombination event, then returns
<cite>-DBL_MAX</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="msprime.log_mutation_likelihood">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">log_mutation_likelihood</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ts</span></em>, <em class="sig-param"><span class="n">mutation_rate</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.log_mutation_likelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the unnormalised log probability of the stored pattern of mutations
on the stored tree sequence, assuming infinite sites mutation. In particular,
each stored site must only contain a single mutation. The omitted normalising
constant depends on the pattern of mutations, but not on the tree sequence or
the mutation rate.</p>
<p>The function first computes the probability of the overall number of mutations
<span class="math notranslate nohighlight">\(M\)</span> from the Poisson probability mass function</p>
<div class="math notranslate nohighlight">
\[e^{-T \mu / 2} \frac{(T \mu / 2)^M}{M!},\]</div>
<p>where <span class="math notranslate nohighlight">\(T\)</span> is the total area of ancestral material in the tree sequence
stored in units of generations, and <span class="math notranslate nohighlight">\(\mu\)</span> is the per-site, per-generation
mutation probability. Each mutation then contributes an individual factor of
<span class="math notranslate nohighlight">\(l / T\)</span>, where <span class="math notranslate nohighlight">\(l\)</span> is the total branch length on which the mutation
could have arisen while appearing on all of the required lineages, again stored
in generations.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If a tree at the site of a mutation contains unary nodes, then <span class="math notranslate nohighlight">\(l\)</span> could
span more than one edge. In particular, we do not constrain mutations to take
place on the edge directly above the node on which they have been recorded,
but rather on any edge which would yield the same configuration of SNPs at the
leaves of the tree sequence.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ts</strong> (<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><em>tskit.TreeSequence</em></a>) – The tree sequence object with mutations.</p></li>
<li><p><strong>mutation_rate</strong> (<em>float</em>) – The per-site, per-generation mutation probablity.
Must be non-negative.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The unnormalised log probability of the observed SNPs given the tree
sequence. If the mutation rate is set to zero and the tree sequence contains
at least one mutation, then returns <cite>-float(“inf”)</cite>.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="id3">
<h2>Mutations<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="msprime.sim_mutations">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">sim_mutations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tree_sequence</span></em>, <em class="sig-param"><span class="n">rate</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">random_seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start_time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end_time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">discrete_genome</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">kept_mutations_before_end_time</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.sim_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates mutations on the specified ancestry and returns the resulting
<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>. Mutations are generated at the specified rate
per unit of sequence length, per generation. By default, mutations are
generated at discrete sites along the genome and multiple mutations
can occur at any given site. A continuous sequence, infinite-sites model
can also be specified by setting the <code class="docutils literal notranslate"><span class="pre">discrete_genome</span></code> parameter to
False.</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">model</span></code> parameter is specified, this determines the model under
which mutations are generated. The default mutation model is
<a class="reference internal" href="#msprime.JC69MutationModel" title="msprime.JC69MutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.JC69MutationModel</span></code></a> a symmetrical mutation model among
the ACGT alleles. See <a class="reference internal" href="mutations.html#sec-mutations-models"><span class="std std-ref">Models</span></a> for details of available models.</p>
<p>If a random seed is specified, this is used to seed the random number
generator. If the same seed is specified and all other parameters are equal
then the same mutations will be generated. If no random seed is specified
then one is generated automatically.</p>
<p>By default, sites and mutations in the input tree sequence are
discarded. If the <code class="docutils literal notranslate"><span class="pre">keep</span></code> parameter is true, however, <em>additional</em>
mutations are simulated. Under the infinite sites mutation model, all new
mutations generated will occur at distinct positions from each other and
from any existing mutations (by rejection sampling). Furthermore, if sites
are discrete, trying to simulate mutations at time periods that are older
than mutations kept from the original tree sequence is an error, because
this would create an extra transition (from the new allele to the old
one below it) that may be incorrect according to the model of mutation.
Under a state-independent mutation model, however (e.g., Jukes-Cantor),
there is no problem, and <code class="docutils literal notranslate"><span class="pre">kept_mutations_before_end_time=True</span></code> may be
set to allow adding new mutations around or above existing ones.</p>
<p>The time interval over which mutations can occur may be controlled
using the <code class="docutils literal notranslate"><span class="pre">start_time</span></code> and <code class="docutils literal notranslate"><span class="pre">end_time</span></code> parameters. The <code class="docutils literal notranslate"><span class="pre">start_time</span></code>
defines the lower bound (in time-ago) on this interval and <code class="docutils literal notranslate"><span class="pre">max_time</span></code>
the upper bound. Note that we may have mutations associated with
nodes with time &lt;= <code class="docutils literal notranslate"><span class="pre">start_time</span></code> since mutations store the node at the
bottom (i.e., towards the leaves) of the branch that they occur on.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><em>tskit.TreeSequence</em></a>) – The tree sequence onto which we
wish to throw mutations.</p></li>
<li><p><strong>rate</strong> (<em>float</em>) – The rate of mutation per generation, as either a
single number (for a uniform rate) or as a
<a class="reference internal" href="#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a>. (Default: 0).</p></li>
<li><p><strong>random_seed</strong> (<em>int</em>) – The random seed. If this is <cite>None</cite>, a
random seed will be automatically generated. Valid random
seeds must be between 1 and <span class="math notranslate nohighlight">\(2^{32} - 1\)</span>.</p></li>
<li><p><strong>model</strong> (<em>MutationModel</em>) – The mutation model to use when generating
mutations. This can either be a string (e.g., <code class="docutils literal notranslate"><span class="pre">&quot;jc69&quot;</span></code>) or
an instance of a simulation model class
e.g, <code class="docutils literal notranslate"><span class="pre">msprime.F84MutationModel(kappa=0.5)</span></code>.
If not specified or None, the <a class="reference internal" href="#msprime.BinaryMutationModel" title="msprime.BinaryMutationModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">BinaryMutationModel</span></code></a>
mutation model is used. Please see the
<a class="reference internal" href="mutations.html#sec-mutations-models"><span class="std std-ref">Models</span></a> section for more details
on specifying mutation models.</p></li>
<li><p><strong>keep</strong> (<em>bool</em>) – Whether to keep existing mutations (default: False).</p></li>
<li><p><strong>start_time</strong> (<em>float</em>) – The minimum time ago at which a mutation can
occur. (Default: no restriction.)</p></li>
<li><p><strong>end_time</strong> (<em>float</em>) – The maximum time ago at which a mutation can occur
(Default: no restriction).</p></li>
<li><p><strong>discrete_genome</strong> (<em>bool</em>) – Whether to generate mutations at only integer positions
along the genome (Default=True).</p></li>
<li><p><strong>kept_mutations_before_end_time</strong> (<em>bool</em>) – Whether to allow mutations to be added
ancestrally to existing (kept) mutations. This flag has no effect
if either keep or discrete_genome are False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> object resulting from overlaying
mutations on the input tree sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="logging.html" title="previous page">Logging</a>
    <a class='right-next' id="next-link" href="development.html" title="next page">Developer documentation</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Tskit Developers<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="_static/js/index.d3f166471bb80abb5163.js"></script>


    
  </body>
</html>