
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Legacy (version 0.x) APIs &#8212; Msprime manual</title>
    
  <link rel="stylesheet" href="_static/css/index.f658d18f9b420779cfdf24aa0a7e2d77.css">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.d3f166471bb80abb5163.js">

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Changelog" href="changelog.html" />
    <link rel="prev" title="Developer documentation" href="development.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="index.html">
  
  
  <h1 class="site-logo" id="site-title">Msprime manual</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Introduction
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="quickstart.html">
   Quickstart
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Using msprime
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="simulations.html">
   Running simulations
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="ancestry.html">
     Ancestry
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="demography.html">
     Demography
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mutations.html">
     Mutations
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="utilities.html">
   Utilities
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="likelihoods.html">
     Computing likelihoods
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="rate_maps.html">
     Rate Maps
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="cli.html">
     Command line interface
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="logging.html">
     Logging
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="api.html">
   API Reference
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="development.html">
   Developer documentation
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Legacy (version 0.x) APIs
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="changelog.html">
   Changelog
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="CITATION.html">
   Citing msprime
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   PORTING IN PROGRESS: Tutorial
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Manual for msprime __MSPRIME_VERSION__
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/legacy.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/tskit-dev/msprime"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/tskit-dev/msprime/issues/new?title=Issue%20on%20page%20%2Flegacy.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/tskit-dev/msprime/edit/main/docs/legacy.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#upgrading-code">
   Upgrading code
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#backwards-compatibility">
   Backwards compatibility
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ancestry">
   Ancestry
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#demography">
   Demography
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mutations">
   Mutations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#utilities">
   Utilities
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#api-reference">
   API Reference
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="legacy-version-0-x-apis">
<h1>Legacy (version 0.x) APIs<a class="headerlink" href="#legacy-version-0-x-apis" title="Permalink to this headline">¶</a></h1>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This page is under construction, and needs work to ensure
that the target audiences are able to find the content they need.</p>
</div>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>An overview of why we added the new APIs, what they are
and some reassurance that old code will continue to work.</p>
</div>
<p>With a few exceptions, you <strong>should not need to change your code</strong> and
it should remain working indefinitely.</p>
<div class="section" id="upgrading-code">
<h2>Upgrading code<a class="headerlink" href="#upgrading-code" title="Permalink to this headline">¶</a></h2>
<p>This section is to help 0.x users of the get up to speed quickly, summarising the new
APIs and their main differences to what you are used to.</p>
<p>The main change is that there are two new functions, <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> and
<a class="reference internal" href="api.html#msprime.sim_mutations" title="msprime.sim_mutations"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_mutations()</span></code></a> which correspond to the 0.x functions <a class="reference internal" href="#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>
and <a class="reference internal" href="#msprime.mutate" title="msprime.mutate"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate()</span></code></a>. The 0.x functions are <strong>deprecated</strong> but <strong>will continue
to be supported indefinitely</strong>.</p>
</div>
<div class="section" id="backwards-compatibility">
<h2>Backwards compatibility<a class="headerlink" href="#backwards-compatibility" title="Permalink to this headline">¶</a></h2>
<p>All existing simulations should work as before, <em>except</em> for simulations relying on
the detailed properties of RecombinationMaps. If your code uses the <code class="docutils literal notranslate"><span class="pre">num_loci</span></code>
property, then it may need to be updated. The reason for this is that <code class="docutils literal notranslate"><span class="pre">msprime</span></code>
has changed to simulate directly in physical coordinates internally (which has
greatly simplified the code and solved many thorny issues) and this is fundamentally
incompatible with the approach taken in 0.x. In the vast majority of cases, this
will have no effect.</p>
<p>If you are using <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> to simulate a discrete genome, it may be simplest to
convert your code to use the new <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> method. If you were following
a recipe to simulate multiple chromosomes under the DTWF model, please see
the <a class="reference internal" href="ancestry.html#sec-ancestry-multiple-chromosomes"><span class="std std-ref">updated recipe</span></a>.</p>
</div>
<div class="section" id="ancestry">
<h2>Ancestry<a class="headerlink" href="#ancestry" title="Permalink to this headline">¶</a></h2>
<p>The new <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> function replaces the 0.x <a class="reference internal" href="#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>
function and is very similar. There are some important differences though:</p>
<ul>
<li><p>The {}<code class="docutils literal notranslate"><span class="pre">samples</span></code> argument now refers to the <strong>number of individuals</strong>
rather than <strong>the number of nodes</strong> (i.e. monoploid genomes).
Because the default <a class="reference internal" href="ancestry.html#sec-ancestry-ploidy"><span class="std std-ref">ploidy</span></a>
is 2 (see the next point) the upshot is that <code class="docutils literal notranslate"><span class="pre">sim_ancestry(2)</span></code> will
result in a tree sequence with <em>four</em> sample nodes, not two. (It is
possible to override this behaviour using the list of <a class="reference internal" href="api.html#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a>
objects argument to <code class="docutils literal notranslate"><span class="pre">samples</span></code>.)</p></li>
<li><p>There is now a <a class="reference internal" href="ancestry.html#sec-ancestry-ploidy"><span class="std std-ref">Ploidy</span></a> argument, which has
two effects:</p>
<ol class="simple">
<li><p>Sets the default number of sample nodes per <em>individual</em></p></li>
<li><p>Changes the timescale of the coalescent process (TODO link to a section
that explains this effect.) By default <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> is 2 and
time is scaled scaled in units of 4N generations, which is the same as
msprime 0.x.</p></li>
</ol>
</li>
<li><p>Rather than two arguments <code class="docutils literal notranslate"><span class="pre">num_samples</span></code> and <code class="docutils literal notranslate"><span class="pre">samples</span></code>, the
<a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> function has a single argument <code class="docutils literal notranslate"><span class="pre">samples</span></code> which
has different behaviour depending on the type of arguments provided.
See <a class="reference internal" href="ancestry.html#sec-ancestry-samples"><span class="std std-ref">Specifying samples</span></a> for details.</p>
<p>Note in particular that a list of <code class="docutils literal notranslate"><span class="pre">Sample</span></code> objects is <strong>not</strong> supported.</p>
</li>
<li><p>Similarly, there is now one argument <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> which can
be either a single value or a <a class="reference internal" href="api.html#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a> object. Note that the
0.x <a class="reference internal" href="#msprime.RecombinationMap" title="msprime.RecombinationMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RecombinationMap</span></code></a> is deprecated and not supported as input
to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>. See <a class="reference internal" href="ancestry.html#sec-ancestry-recombination"><span class="std std-ref">Recombination</span></a> for more
details.</p></li>
<li><p>Simulations are peformed on a <strong>discrete</strong> genome by default. To get the
0.x behaviour of a continuous genome, set <code class="docutils literal notranslate"><span class="pre">discrete_genome=False</span></code>.
See <a class="reference internal" href="ancestry.html#sec-ancestry-discrete-genome"><span class="std std-ref">Discrete or continuous?</span></a> for more details.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">from_ts</span></code> argument used has been renamed to <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> and
accepts either a <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TableCollection" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TableCollection</span></code></a> or <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>
argument. See <a class="reference internal" href="ancestry.html#sec-ancestry-initial-state"><span class="std std-ref">Specifying the initial state</span></a> for details.</p></li>
<li><p>There is <strong>no</strong> <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> argument to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>: use
<a class="reference internal" href="api.html#msprime.sim_mutations" title="msprime.sim_mutations"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_mutations()</span></code></a> instead.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">population_configurations</span></code>, <code class="docutils literal notranslate"><span class="pre">migration_matrix</span></code> and <code class="docutils literal notranslate"><span class="pre">demographic_events</span></code>
arguments have been replace with a single argument <code class="docutils literal notranslate"><span class="pre">demography</span></code>, which must take
a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> instance. (See the next section for more details.)</p></li>
</ul>
</div>
<div class="section" id="demography">
<h2>Demography<a class="headerlink" href="#demography" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>A new <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object has been added for version 1.0 which
encapsulates the functionality needed to define and debug demographic models
in msprime. Demographic models can only be specified to <code class="docutils literal notranslate"><span class="pre">sim_ancestry</span></code>
using an instance of this class.</p></li>
<li><p>It is easy to create a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> from the 0.x
<code class="docutils literal notranslate"><span class="pre">population_configurations</span></code>, <code class="docutils literal notranslate"><span class="pre">migration_matrix</span></code> and <code class="docutils literal notranslate"><span class="pre">demographic_events</span></code>
values using the <a class="reference internal" href="api.html#msprime.Demography.from_old_style" title="msprime.Demography.from_old_style"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_old_style()</span></code></a> method.</p></li>
<li><p>The <a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> class should no longer be instantiated
directly; instead use the <a class="reference internal" href="api.html#msprime.Demography.debug" title="msprime.Demography.debug"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.debug()</span></code></a> method.</p></li>
</ul>
</div>
<div class="section" id="mutations">
<h2>Mutations<a class="headerlink" href="#mutations" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>For symmetry with the <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> function, there is now a <a class="reference internal" href="api.html#msprime.sim_mutations" title="msprime.sim_mutations"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_mutations()</span></code></a>
function. The 0.x <a class="reference internal" href="#msprime.mutate" title="msprime.mutate"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate()</span></code></a> function is <strong>deprecated</strong>.</p></li>
<li><p>The <a class="reference internal" href="api.html#msprime.sim_mutations" title="msprime.sim_mutations"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_mutations()</span></code></a> function works on a <strong>discrete</strong> genome by default.</p></li>
<li><p>There are now also many new mutation models supported by :func:.sim_mutations;
see <a class="reference internal" href="mutations.html#sec-mutations"><span class="std std-ref">Mutations</span></a> for details. These are <em>not</em> supported in the deprecated
<a class="reference internal" href="#msprime.mutate" title="msprime.mutate"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate()</span></code></a> function.</p></li>
<li><p>The simulated mutations now have a simulated <code class="docutils literal notranslate"><span class="pre">time</span></code> value, which specifies the
precise time that the mutation occurred. Note that this value is also provided in the
returned tables for the deprecated <code class="docutils literal notranslate"><span class="pre">simulate()</span></code> and <code class="docutils literal notranslate"><span class="pre">mutate()</span></code> functions,
which may lead to some compatibility issues. (We considered removing the simulated
mutation times for these 0.x functions for strict compatibility, but this would
have broken any code using the <code class="docutils literal notranslate"><span class="pre">keep</span></code> option in mutate.)</p></li>
</ul>
</div>
<div class="section" id="utilities">
<h2>Utilities<a class="headerlink" href="#utilities" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>The 0.x class <a class="reference internal" href="#msprime.RecombinationMap" title="msprime.RecombinationMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RecombinationMap</span></code></a> has been <strong>deprecated</strong> in favour of the new
<a class="reference internal" href="api.html#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a>. This was to (a) generalise the interface to accomodate varying
rates of mutation and gene conversion along the genome; and (b) convert to a
more modern numpy-based API.</p></li>
</ul>
</div>
<div class="section" id="api-reference">
<h2>API Reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="msprime.PopulationConfiguration">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">PopulationConfiguration</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">initial_size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.PopulationConfiguration" title="Permalink to this definition">¶</a></dt>
<dd><p>The initial configuration of a population (or deme) in a simulation.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This is deprecated. Document as such.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_size</strong> (<em>int</em>) – The number of initial samples that are drawn
from this population.</p></li>
<li><p><strong>initial_size</strong> (<em>float</em>) – The absolute size of the population at time
zero. Defaults to the reference population size <span class="math notranslate nohighlight">\(N_e\)</span>.</p></li>
<li><p><strong>growth_rate</strong> (<em>float</em>) – The forwards-time exponential growth rate of the
population per generation. Growth rates can be negative. This is zero for a
constant population size, and positive for a population that has been
growing. Defaults to 0.</p></li>
<li><p><strong>metadata</strong> (<em>dict</em>) – A JSON-encodable dictionary of metadata to associate
with the corresponding Population in the output tree sequence.
If not specified or None, no metadata is stored (i.e., an empty bytes array).
Note that this metadata is ignored when using the <code class="docutils literal notranslate"><span class="pre">from_ts</span></code> argument to
<a class="reference internal" href="#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>, as the population definitions in the tree sequence that
is used as the starting point take precedence.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.RecombinationMap">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">RecombinationMap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">positions</span></em>, <em class="sig-param"><span class="n">rates</span></em>, <em class="sig-param"><span class="n">num_loci</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">map_start</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RecombinationMap" title="Permalink to this definition">¶</a></dt>
<dd><p>A RecombinationMap represents the changing rates of recombination
along a chromosome. This is defined via two lists of numbers:
<code class="docutils literal notranslate"><span class="pre">positions</span></code> and <code class="docutils literal notranslate"><span class="pre">rates</span></code>, which must be of the same length.
Given an index j in these lists, the rate of recombination
per base per generation is <code class="docutils literal notranslate"><span class="pre">rates[j]</span></code> over the interval
<code class="docutils literal notranslate"><span class="pre">positions[j]</span></code> to <code class="docutils literal notranslate"><span class="pre">positions[j</span> <span class="pre">+</span> <span class="pre">1]</span></code>. Consequently, the first
position must be zero, and by convention the last rate value
is also required to be zero (although it is not used).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> parameter is deprecated.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>positions</strong> (<em>list</em>) – The positions (in bases) denoting the
distinct intervals where recombination rates change. These can
be floating point values.</p></li>
<li><p><strong>rates</strong> (<em>list</em>) – The list of rates corresponding to the supplied
<code class="docutils literal notranslate"><span class="pre">positions</span></code>. Recombination rates are specified per base,
per generation.</p></li>
<li><p><strong>num_loci</strong> (<em>int</em>) – <strong>This parameter is deprecated</strong>.
The maximum number of non-recombining loci
in the underlying simulation. By default this is set to
the largest possible value, allowing the maximum resolution
in the recombination process. However, for a finite sites
model this can be set to smaller values.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="msprime.RecombinationMap.get_total_recombination_rate">
<code class="sig-name descname">get_total_recombination_rate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RecombinationMap.get_total_recombination_rate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the effective recombination rate for this genetic map.
This is the weighted mean of the rates across all intervals.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RecombinationMap.mean_recombination_rate">
<em class="property">property </em><code class="sig-name descname">mean_recombination_rate</code><a class="headerlink" href="#msprime.RecombinationMap.mean_recombination_rate" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the weighted mean recombination rate
across all windows of the entire recombination map.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.RecombinationMap.read_hapmap">
<em class="property">classmethod </em><code class="sig-name descname">read_hapmap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RecombinationMap.read_hapmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the specified file in HapMap format.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method is deprecated, use the module-level
<a class="reference internal" href="#msprime.RecombinationMap.read_hapmap" title="msprime.RecombinationMap.read_hapmap"><code class="xref py py-func docutils literal notranslate"><span class="pre">read_hapmap()</span></code></a> function instead.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>str</em>) – The name of the file to be parsed. This may be
in plain text or gzipped plain text.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A RecombinationMap object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.RecombinationMap.uniform_map">
<em class="property">classmethod </em><code class="sig-name descname">uniform_map</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">length</span></em>, <em class="sig-param"><span class="n">rate</span></em>, <em class="sig-param"><span class="n">num_loci</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.RecombinationMap.uniform_map" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#msprime.RecombinationMap" title="msprime.RecombinationMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RecombinationMap</span></code></a> instance in which the recombination
rate is constant over a chromosome of the specified length.
The legacy <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> option is no longer supported and should not be used.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>length</strong> (<em>float</em>) – The length of the chromosome.</p></li>
<li><p><strong>rate</strong> (<em>float</em>) – The rate of recombination per unit of sequence length
along this chromosome.</p></li>
<li><p><strong>num_loci</strong> (<em>int</em>) – This parameter is no longer supported.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="msprime.simulate">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">simulate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#msprime.simulate" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates the coalescent with recombination under the specified model
parameters and returns the resulting <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>. Note that
Ne is the effective diploid population size (so the effective number
of genomes in the population is 2*Ne), but <code class="docutils literal notranslate"><span class="pre">sample_size</span></code> is the
number of (monoploid) genomes sampled.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_size</strong> (<em>int</em>) – The number of sampled monoploid genomes.  If not
specified or None, this defaults to the sum of the subpopulation sample
sizes. Either <code class="docutils literal notranslate"><span class="pre">sample_size</span></code>, <code class="docutils literal notranslate"><span class="pre">population_configurations</span></code> or
<code class="docutils literal notranslate"><span class="pre">samples</span></code> must be specified.</p></li>
<li><p><strong>Ne</strong> (<em>float</em>) – The effective (diploid) population size. This defaults to
1 if not specified.</p></li>
<li><p><strong>length</strong> (<em>float</em>) – The length of the simulated region in bases.
This parameter cannot be used along with <code class="docutils literal notranslate"><span class="pre">recombination_map</span></code>.
Defaults to 1 if not specified.</p></li>
<li><p><strong>recombination_rate</strong> (<em>float</em>) – The rate of recombination per base
per generation. This parameter cannot be used along with
<code class="docutils literal notranslate"><span class="pre">recombination_map</span></code>. Defaults to 0 if not specified.</p></li>
<li><p><strong>recombination_map</strong> (<a class="reference internal" href="#msprime.RecombinationMap" title="msprime.RecombinationMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RecombinationMap</span></code></a>) – The map
describing the changing rates of recombination along the simulated
chromosome. This parameter cannot be used along with the
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> or <code class="docutils literal notranslate"><span class="pre">length</span></code> parameters, as these
values are encoded within the map. Defaults to a uniform rate as
described in the <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> parameter if not specified.</p></li>
<li><p><strong>mutation_rate</strong> (<em>float</em>) – The rate of infinite sites
mutations per unit of sequence length per generation.
If not specified, no mutations are generated. This option only
allows for infinite sites mutations with a binary (i.e., 0/1)
alphabet. For more control over the mutational process, please
use the <a class="reference internal" href="#msprime.mutate" title="msprime.mutate"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate()</span></code></a> function.</p></li>
<li><p><strong>population_configurations</strong> (<em>list</em><em> or </em><em>None.</em>) – The list of
<a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a> instances describing the
sampling configuration, relative sizes and growth rates of
the populations to be simulated. If this is not specified,
a single population with a sample of size <code class="docutils literal notranslate"><span class="pre">sample_size</span></code>
is assumed.</p></li>
<li><p><strong>migration_matrix</strong> (<em>list</em>) – The matrix describing the rates of migration
between all pairs of populations. If <span class="math notranslate nohighlight">\(N\)</span> populations are defined
in the <code class="docutils literal notranslate"><span class="pre">population_configurations</span></code> parameter, then the migration
matrix must be an <span class="math notranslate nohighlight">\(N \times N\)</span> matrix with 0 on the diagonal,
consisting of <span class="math notranslate nohighlight">\(N\)</span> lists of length <span class="math notranslate nohighlight">\(N\)</span> or an <span class="math notranslate nohighlight">\(N
\times N\)</span> numpy array. The <span class="math notranslate nohighlight">\([j, k]^{th}\)</span> element of the
migration matrix gives the expected number of migrants moving from
population <span class="math notranslate nohighlight">\(k\)</span> to population <span class="math notranslate nohighlight">\(j\)</span> per generation, divided by
the size of population <span class="math notranslate nohighlight">\(j\)</span>.  When simulating from the
discrete-time Wright-Fisher model (<code class="docutils literal notranslate"><span class="pre">model</span> <span class="pre">=</span> <span class="pre">&quot;dtwf&quot;</span></code>), the row sums of
the migration matrix must not exceed 1. There are no sum constraints for
migration rates in continuous-time models.</p></li>
<li><p><strong>demographic_events</strong> (<em>list</em>) – The list of demographic events to
simulate. Demographic events describe changes to the populations
in the past. Events should be supplied in non-decreasing
order of time in the past. Events with the same time value will be
applied sequentially in the order that they were supplied before the
simulation algorithm continues with the next time step.</p></li>
<li><p><strong>samples</strong> (<em>list</em>) – The list specifying the location and time of
all samples. This parameter may be used to specify historical
samples, and cannot be used in conjunction with the <code class="docutils literal notranslate"><span class="pre">sample_size</span></code>
parameter. Each sample is a (<code class="docutils literal notranslate"><span class="pre">population</span></code>, <code class="docutils literal notranslate"><span class="pre">time</span></code>) pair
such that the sample in position <code class="docutils literal notranslate"><span class="pre">j</span></code> in the list of samples
is drawn in the specified population at the specfied time. Time
is measured in generations ago, as elsewhere.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em>) – The random seed. If this is <cite>None</cite>, a
random seed will be automatically generated. Valid random
seeds must be between 1 and <span class="math notranslate nohighlight">\(2^{32} - 1\)</span>.</p></li>
<li><p><strong>num_replicates</strong> (<em>int</em>) – The number of replicates of the specified
parameters to simulate. If this is not specified or None,
no replication is performed and a <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> object
returned. If <code class="xref py py-obj docutils literal notranslate"><span class="pre">num_replicates</span></code> is provided, the specified
number of replicates is performed, and an iterator over the
resulting <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> objects returned.</p></li>
<li><p><strong>from_ts</strong> (<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><em>tskit.TreeSequence</em></a>) – If specified, initialise the simulation
from the root segments of this tree sequence and return the
completed tree sequence. Please see <a class="reference internal" href="ancestry.html#sec-ancestry-initial-state"><span class="std std-ref">here</span></a> for details on the required properties
of this tree sequence and its interactions with other parameters.
(Default: None).</p></li>
<li><p><strong>start_time</strong> (<em>float</em>) – If specified, set the initial time that the
simulation starts to this value. If not specified, the start
time is zero if performing a simulation of a set of samples,
or is the time of the oldest node if simulating from an
existing tree sequence (see the <code class="docutils literal notranslate"><span class="pre">from_ts</span></code> parameter).</p></li>
<li><p><strong>end_time</strong> (<em>float</em>) – If specified, terminate the simulation at the
specified time. In the returned tree sequence, all rootward paths from
samples with time &lt; end_time will end in a node with one child with
time equal to end_time. Sample nodes with time &gt;= end_time will
also be present in the output tree sequence. If not specified or <code class="docutils literal notranslate"><span class="pre">None</span></code>,
run the simulation until all samples have an MRCA at all positions in
the genome.</p></li>
<li><p><strong>record_full_arg</strong> (<em>bool</em>) – If True, record all intermediate nodes
arising from common ancestor and recombination events in the output
tree sequence. This will result in unary nodes (i.e., nodes in marginal
trees that have only one child). Defaults to False.</p></li>
<li><p><strong>model</strong> (<em>str</em><em> or </em><em>simulation model instance</em>) – The simulation model to use.
This can either be a string (e.g., <code class="docutils literal notranslate"><span class="pre">&quot;smc_prime&quot;</span></code>) or an instance of
a simulation model class (e.g, <code class="docutils literal notranslate"><span class="pre">msprime.DiscreteTimeWrightFisher()</span></code>.
Please see the <a class="reference internal" href="ancestry.html#sec-ancestry-models"><span class="std std-ref">Models</span></a> section for more details
on specifying ancestry models.</p></li>
<li><p><strong>record_provenance</strong> (<em>bool</em>) – If True, record all configuration and parameters
required to recreate the tree sequence. These can be accessed
via <code class="docutils literal notranslate"><span class="pre">TreeSequence.provenances()</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> object representing the results
of the simulation if no replication is performed, or an
iterator over the independent replicates simulated if the
<code class="xref py py-obj docutils literal notranslate"><span class="pre">num_replicates</span></code> parameter has been used.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> or an iterator over
<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> replicates.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.SimulationModelChange">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">SimulationModelChange</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>Union<span class="p">[</span>str<span class="p">, </span>msprime.ancestry.AncestryModel<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.SimulationModelChange" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated 0.x way to describe an <a class="reference internal" href="api.html#msprime.AncestryModelChange" title="msprime.AncestryModelChange"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestryModelChange</span></code></a>.</p>
</dd></dl>

<dl class="py function">
<dt id="msprime.mutate">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">mutate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tree_sequence</span></em>, <em class="sig-param"><span class="n">rate</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">random_seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">start_time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end_time</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#msprime.mutate" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates mutations on the specified ancestry and returns the resulting
<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>. Mutations are generated at the specified rate in
measured generations. Mutations are generated under the infinite sites
model, and so the rate of new mutations is per unit of sequence length per
generation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is deprecated and we recommend new code use
the <a class="reference internal" href="api.html#msprime.sim_mutations" title="msprime.sim_mutations"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_mutations()</span></code></a> function instead, which is similar
but offers more functionality and defaults to simulating mutations
on a discrete genome. This function will be supported indefinitely,
however.</p>
</div>
<p>If a random seed is specified, this is used to seed the random number
generator. If the same seed is specified and all other parameters are equal
then the same mutations will be generated. If no random seed is specified
then one is generated automatically.</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">model</span></code> parameter is specified, this determines the model under
which mutations are generated. By default mutations from the
infinite sites model with a binary alphabet are generated.</p>
<p>By default, sites and mutations in the parameter tree sequence are
discarded. If the <code class="docutils literal notranslate"><span class="pre">keep</span></code> parameter is true, however, <em>additional</em>
mutations are simulated. Under the infinite sites mutation model, all new
mutations generated will occur at distinct positions from each other and
from any existing mutations (by rejection sampling).</p>
<p>The time interval over which mutations can occur may be controlled
using the <code class="docutils literal notranslate"><span class="pre">start_time</span></code> and <code class="docutils literal notranslate"><span class="pre">end_time</span></code> parameters. The <code class="docutils literal notranslate"><span class="pre">start_time</span></code>
defines the lower bound (in time-ago) on this interval and <code class="docutils literal notranslate"><span class="pre">max_time</span></code>
the upper bound. Note that we may have mutations associated with
nodes with time &lt;= <code class="docutils literal notranslate"><span class="pre">start_time</span></code> since mutations store the node at the
bottom (i.e., towards the leaves) of the branch that they occur on.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><em>tskit.TreeSequence</em></a>) – The tree sequence onto which we
wish to throw mutations.</p></li>
<li><p><strong>rate</strong> (<em>float</em>) – The rate of mutation per generation. (Default: 0).</p></li>
<li><p><strong>random_seed</strong> (<em>int</em>) – The random seed. If this is <cite>None</cite>, a
random seed will be automatically generated. Valid random
seeds must be between 1 and <span class="math notranslate nohighlight">\(2^{32} - 1\)</span>.</p></li>
<li><p><strong>model</strong> (<em>MutationModel</em>) – The mutation model to use when generating
mutations. If not specified or None, the <code class="xref py py-class docutils literal notranslate"><span class="pre">InfiniteSites</span></code>
mutation model is used.</p></li>
<li><p><strong>keep</strong> (<em>bool</em>) – Whether to keep existing mutations (default: False).</p></li>
<li><p><strong>start_time</strong> (<em>float</em>) – The minimum time at which a mutation can
occur. (Default: no restriction.)</p></li>
<li><p><strong>end_time</strong> (<em>float</em>) – The maximum time at which a mutation can occur
(Default: no restriction).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> object  resulting from overlaying
mutations on the input tree sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.readthedocs.io/en/stable/python-api.html#tskit.TreeSequence" title="(in tskit v0.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="development.html" title="previous page">Developer documentation</a>
    <a class='right-next' id="next-link" href="changelog.html" title="next page">Changelog</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Tskit Developers<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="_static/js/index.d3f166471bb80abb5163.js"></script>


    
  </body>
</html>