
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Quickstart &#8212; Msprime manual</title>
    
  <link rel="stylesheet" href="_static/css/index.f658d18f9b420779cfdf24aa0a7e2d77.css">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.d3f166471bb80abb5163.js">

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="prev" title="Introduction" href="intro.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="index.html">
  
  
  <h1 class="site-logo" id="site-title">Msprime manual</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Introduction
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Quickstart
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Using msprime
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="simulations.html">
   Running simulations
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="ancestry.html">
     Ancestry
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="demography.html">
     Demography
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mutations.html">
     Mutations
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="utilities.html">
   Utilities
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="likelihoods.html">
     Computing likelihoods
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="rate_maps.html">
     Rate Maps
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="cli.html">
     Command line interface
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="logging.html">
     Logging
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="api.html">
   API Reference
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="development.html">
   Developer documentation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="legacy.html">
   Legacy (version 0.x) APIs
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="changelog.html">
   Changelog
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="CITATION.html">
   Citing msprime
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   PORTING IN PROGRESS: Tutorial
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Manual for msprime __MSPRIME_VERSION__
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="_sources/quickstart.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/quickstart.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/tskit-dev/msprime"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/tskit-dev/msprime/issues/new?title=Issue%20on%20page%20%2Fquickstart.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/tskit-dev/msprime/edit/main/docs/quickstart.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/tskit-dev/msprime/main?urlpath=tree/docs/quickstart.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ancestry">
   Ancestry
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mutations">
   Mutations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#demography">
   Demography
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="quickstart">
<span id="sec-quickstart"></span><h1>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h1>
<p>This page gives some simple examples of how to use the major features
of <strong class="program">msprime</strong>, with links to more detailed documentation
and tutorial content.</p>
<p>See the <a class="reference internal" href="installation.html#sec-installation"><span class="std std-ref">Installation</span></a> page for instructions on installing
<strong class="program">msprime</strong> (short version: <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">msprime</span></code> or
<code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">conda-forge</span> <span class="pre">msprime</span></code> will work for most users).</p>
<div class="section" id="ancestry">
<span id="sec-quickstart-ancestry"></span><h2>Ancestry<a class="headerlink" href="#ancestry" title="Permalink to this headline">¶</a></h2>
<p>Msprime simulates ancestral histories for a set of sample genomes
using backwards-in-time population genetic models.
Here we run a simple simulation of a short recombining sequence under
human-like parameters:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>    <span class="kn">import</span> <span class="nn">msprime</span>
    <span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">SVG</span>

    <span class="c1"># Simulate an ancestral history for 3 diploid samples under the coalescent</span>
    <span class="c1"># with recombination on a 5kb region with human-like parameters.</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
        <span class="n">samples</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> 
        <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> 
        <span class="n">sequence_length</span><span class="o">=</span><span class="mi">5_000</span><span class="p">,</span>
        <span class="n">population_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span> 
        <span class="n">random_seed</span><span class="o">=</span><span class="mi">123456</span><span class="p">)</span>
    <span class="c1"># Visualise the simulated ancestral history.</span>
    <span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/quickstart_1_0.svg" src="_images/quickstart_1_0.svg" /></div>
</div>
<p>In this example we simulate the ancestral history of three diploid
individuals (see <a class="reference internal" href="ancestry.html#sec-ancestry-samples"><span class="std std-ref">Specifying samples</span></a> and <a class="reference internal" href="ancestry.html#sec-ancestry-ploidy"><span class="std std-ref">Ploidy</span></a>)
for a 5kb sequence with a
<a class="reference external" href="https://en.wikipedia.org/wiki/Genetic_recombination">recombination</a>
rate of <span class="math notranslate nohighlight">\(10^{-8}\)</span>
(see <a class="reference internal" href="ancestry.html#sec-ancestry-genome-properties"><span class="std std-ref">Genome properties</span></a>)
from a population with a constant size of 10,000 (see
the <a class="reference internal" href="#sec-quickstart-demography"><span class="std std-ref">Demography</span></a> section below)
under the default
<a class="reference external" href="https://en.wikipedia.org/wiki/Coalescent_theory">coalescent</a>
ancestry model (see the <a class="reference internal" href="ancestry.html#sec-ancestry-models"><span class="std std-ref">Models</span></a> for details on
other available models).
To ensure that
the output of this example is predictable, we set a random seed
(see <a class="reference internal" href="ancestry.html#sec-ancestry-random-seed"><span class="std std-ref">Random seeds</span></a>).</p>
<p>When recombination is present, the ancestry of a sample of DNA sequences
cannot be represented by a single genealogical tree relating the
samples to their genetic ancestors; there is instead
a <em>sequence</em> of highly correlated trees along the genome.
The result of our simulation is therefore a <a class="reference external" href="https://tskit.dev">tree sequence</a>
object from the <a class="reference external" href="https://tskit.readthedocs.io/en/stable/introduction.html#sec-introduction" title="(in tskit v0.3)"><span class="xref std std-ref">tskit</span></a> library,
which provides a rich suite of operations for
analysing these genealogical histories: see the
<a class="reference external" href="https://tskit.dev/tutorials/getting_started.html#sec-tskit-getting-started" title="(in Python)"><span>Getting started with tskit</span></a> tutorial for help.
In this example we show a visualisation
of the four different trees along the 5kb region
(see the <a class="reference external" href="https://tskit.dev/tutorials/viz.html#sec-tskit-viz" title="(in Python)"><span>Visualisation</span></a> tutorial for more
examples).  Because we have specified three diploid sample
<em>individuals</em>, each of these trees has 6 “sample” nodes
(the “leaves” or “tips”), because each diploid individual
has two monoploid genomes (see <a class="reference internal" href="ancestry.html#sec-ancestry-samples"><span class="std std-ref">Specifying samples</span></a>).</p>
<p>See the <a class="reference internal" href="ancestry.html#sec-ancestry"><span class="std std-ref">Ancestry</span></a> section for more details on
ancestry simulations.</p>
</div>
<div class="section" id="mutations">
<h2>Mutations<a class="headerlink" href="#mutations" title="Permalink to this headline">¶</a></h2>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This is a WIP</p>
</div>
<p>The <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> function generates a simulated ancestral
history for some samples.
If we want <a class="reference external" href="https://en.wikipedia.org/wiki/Genome">genome sequence</a>
we must also simulate some
<a class="reference external" href="https://en.wikipedia.org/wiki/Mutation">mutations</a> on these trees.
However, it’s important to note that it’s not always necessary to
simulate mutations in order to use the simulations; often, it’s
better <em>not to</em>; see the</p>
<p>tutorial for more information.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mutated_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">54321</span><span class="p">)</span>
<span class="n">mutated_ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">sites</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>position</th><th>ancestral_state</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>90.00000000</td><td>T</td><td></td></tr>
<tr><td>1</td><td>333.00000000</td><td>G</td><td></td></tr>
<tr><td>2</td><td>819.00000000</td><td>T</td><td></td></tr>
<tr><td>3</td><td>3204.00000000</td><td>A</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mutated_ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">mutations</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>site</th><th>node</th><th>time</th><th>derived_state</th><th>parent</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>0</td><td>10</td><td>12191.4064954057</td><td>G</td><td>-1</td><td></td></tr>
<tr><td>1</td><td>1</td><td>10</td><td>44173.26473294006</td><td>C</td><td>-1</td><td></td></tr>
<tr><td>2</td><td>2</td><td>9</td><td>9597.219523507081</td><td>G</td><td>-1</td><td></td></tr>
<tr><td>3</td><td>3</td><td>3</td><td>1158.4291490099815</td><td>C</td><td>-1</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">SVG</span><span class="p">(</span><span class="n">mutated_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/quickstart_5_0.svg" src="_images/quickstart_5_0.svg" /></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">mutated_ts</span><span class="o">.</span><span class="n">variants</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;site&#39;: {&#39;id&#39;: 0, &#39;position&#39;: 90.0, &#39;ancestral_state&#39;: &#39;T&#39;, &#39;mutations&#39;: [{&#39;id&#39;: 0, &#39;site&#39;: 0, &#39;node&#39;: 10, &#39;time&#39;: 12191.4064954057, &#39;derived_state&#39;: &#39;G&#39;, &#39;parent&#39;: -1, &#39;metadata&#39;: b&#39;&#39;}], &#39;metadata&#39;: b&#39;&#39;}, &#39;alleles&#39;: (&#39;T&#39;, &#39;G&#39;), &#39;has_missing_data&#39;: False, &#39;num_alleles&#39;: 2, &#39;genotypes&#39;: array([1, 1, 1, 0, 1, 1], dtype=int8), &#39;position&#39;: 90.0, &#39;index&#39;: 0}
{&#39;site&#39;: {&#39;id&#39;: 1, &#39;position&#39;: 333.0, &#39;ancestral_state&#39;: &#39;G&#39;, &#39;mutations&#39;: [{&#39;id&#39;: 1, &#39;site&#39;: 1, &#39;node&#39;: 10, &#39;time&#39;: 44173.26473294006, &#39;derived_state&#39;: &#39;C&#39;, &#39;parent&#39;: -1, &#39;metadata&#39;: b&#39;&#39;}], &#39;metadata&#39;: b&#39;&#39;}, &#39;alleles&#39;: (&#39;G&#39;, &#39;C&#39;), &#39;has_missing_data&#39;: False, &#39;num_alleles&#39;: 2, &#39;genotypes&#39;: array([1, 1, 1, 0, 1, 1], dtype=int8), &#39;position&#39;: 333.0, &#39;index&#39;: 1}
{&#39;site&#39;: {&#39;id&#39;: 2, &#39;position&#39;: 819.0, &#39;ancestral_state&#39;: &#39;T&#39;, &#39;mutations&#39;: [{&#39;id&#39;: 2, &#39;site&#39;: 2, &#39;node&#39;: 9, &#39;time&#39;: 9597.219523507081, &#39;derived_state&#39;: &#39;G&#39;, &#39;parent&#39;: -1, &#39;metadata&#39;: b&#39;&#39;}], &#39;metadata&#39;: b&#39;&#39;}, &#39;alleles&#39;: (&#39;T&#39;, &#39;G&#39;), &#39;has_missing_data&#39;: False, &#39;num_alleles&#39;: 2, &#39;genotypes&#39;: array([1, 0, 1, 0, 1, 0], dtype=int8), &#39;position&#39;: 819.0, &#39;index&#39;: 2}
{&#39;site&#39;: {&#39;id&#39;: 3, &#39;position&#39;: 3204.0, &#39;ancestral_state&#39;: &#39;A&#39;, &#39;mutations&#39;: [{&#39;id&#39;: 3, &#39;site&#39;: 3, &#39;node&#39;: 3, &#39;time&#39;: 1158.4291490099815, &#39;derived_state&#39;: &#39;C&#39;, &#39;parent&#39;: -1, &#39;metadata&#39;: b&#39;&#39;}], &#39;metadata&#39;: b&#39;&#39;}, &#39;alleles&#39;: (&#39;A&#39;, &#39;C&#39;), &#39;has_missing_data&#39;: False, &#39;num_alleles&#39;: 2, &#39;genotypes&#39;: array([0, 0, 0, 1, 0, 0], dtype=int8), &#39;position&#39;: 3204.0, &#39;index&#39;: 3}
</pre></div>
</div>
</div>
</div>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Not sure how much detail we want to get into here. The salient points
we want to get across are that mutations are not automatically part of the
simulation, that we only output sites with mutations and we have efficient
ways to work with the results. A lot of this should be pointing to the
“getting started with tskit” tutorial.</p>
</div>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>List of pointers to the relevant sections of the documentation.</p>
</div>
</div>
<div class="section" id="demography">
<span id="sec-quickstart-demography"></span><h2>Demography<a class="headerlink" href="#demography" title="Permalink to this headline">¶</a></h2>
<p>By default ancestry simulations assume an extremely simple
population structure in which a single randomly mating population
of a fixed size exists for all time. For most simulations this
is an unrealistic assumption, and so msprime provides a way
to describe more complex demographic histories.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>    <span class="c1"># Create a 1D stepping stone model of demograpy</span>
    <span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">stepping_stone_model</span><span class="p">([</span><span class="mi">100</span><span class="p">]</span> <span class="o">*</span> <span class="mi">10</span><span class="p">,</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
    <span class="c1"># Take one diploid sample each from the first and last demes</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">9</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>
    <span class="c1"># Simulate an ancestral history for this demography and sample.</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000000000</td><td></td></tr>
<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.00000000000000</td><td></td></tr>
<tr><td>2</td><td>1</td><td>9</td><td>1</td><td>0.00000000000000</td><td></td></tr>
<tr><td>3</td><td>1</td><td>9</td><td>1</td><td>0.00000000000000</td><td></td></tr>
<tr><td>4</td><td>0</td><td>1</td><td>-1</td><td>858.95831903365752</td><td></td></tr>
<tr><td>5</td><td>0</td><td>5</td><td>-1</td><td>1045.59130258085588</td><td></td></tr>
<tr><td>6</td><td>0</td><td>2</td><td>-1</td><td>6799.39277301452785</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Links into more detailed documentation</p>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="intro.html" title="previous page">Introduction</a>
    <a class='right-next' id="next-link" href="installation.html" title="next page">Installation</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Tskit Developers<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="_static/js/index.d3f166471bb80abb5163.js"></script>


    
  </body>
</html>