{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "parental-chase",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "    import msprime\n",
    "    from IPython.display import SVG\n",
    "    # Doing this to make the notebook outputs deterministic. \n",
    "    # DO NOT DO THIS IN YOUR CODE\n",
    "    msprime.core.set_seed_rng_seed(42)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "romance-bryan",
   "metadata": {},
   "source": [
    "(sec_mutations)=\n",
    "\n",
    "# Mutations\n",
    "\n",
    "\n",
    ":::{warning}\n",
    "\n",
    "This documentation is under heavy construction. Please note\n",
    "any outstanding TODOs before opening issues.\n",
    "\n",
    ":::\n",
    "\n",
    "---\n",
    "\n",
    "## Quick reference\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: \n",
    "  These summaries need work. We can say more then \"amino acids\" e.g. for PAM\n",
    "```\n",
    "\n",
    "{func}`.sim_mutations`\n",
    ": Simulate mutations on ancestral topology  \n",
    "\n",
    "**Models**\n",
    "\n",
    "{class}`.BinaryMutationModel`             \n",
    ": 0/1 flip-flopping alleles                 \n",
    "\n",
    "{class}`.JC69MutationModel`               \n",
    ": Jukes & Cantor '69, nucleotides           \n",
    "\n",
    "{class}`.HKYMutationModel`                \n",
    ": Hasegawa, Kishino & Yano '85, nucleotides \n",
    "\n",
    "{class}`.F84MutationModel`                \n",
    ": Felsenstein '84, nucleotides              \n",
    "\n",
    "{class}`.GTRMutationModel`                \n",
    ": general time-reversible, nucleotides      \n",
    "\n",
    "{class}`.BLOSUM62MutationModel` \n",
    ": amino acids                               \n",
    "\n",
    "{class}`.PAMMutationModel`              \n",
    ": amino acids                               \n",
    "\n",
    "{class}`.MatrixMutationModel`             \n",
    ": general finite-state mutations            \n",
    "\n",
    "{class}`.InfiniteAllelesMutationModel`    \n",
    ": a generic infinite-alleles model          \n",
    "\n",
    "{class}`.SLiMMutationModel`               \n",
    ": SLiM compatible mutations                 \n",
    "\n",
    "---\n",
    "\n",
    "```{eval-rst}\n",
    "\n",
    ".. todo:: \n",
    "    This has been copied directly from the api.rst file. Probably\n",
    "    needs some refactoring to use the newer docs arrangements.\n",
    "    Also all the labels need updating.\n",
    "```\n",
    "\n",
    "When running coalescent simulations it's usually most convenient to use the\n",
    "`mutation_rate` argument to the {func}`.simulate` function to throw neutral\n",
    "mutations down on the trees. However, sometimes we wish to throw mutations\n",
    "down on an existing tree sequence: for example, if we want to see the outcome\n",
    "of different random mutational processes on top of a single simulated topology,\n",
    "or if we have obtained the tree sequence from another program and wish to\n",
    "overlay neutral mutations on this tree sequence.\n",
    "\n",
    "\n",
    "(sec_mutations_rate)=\n",
    "\n",
    "## Specifying rates\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: Documentation on the ``rate`` parameter with examples.\n",
    "\n",
    "```\n",
    "\n",
    "(sec_mutations_randomness)=\n",
    "\n",
    "## Controlling randomness\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: Documentation on the ``seed` parameter with examples. Can\n",
    "    link to the randomness section in ancstry.rst. Alternatively we\n",
    "    don't bother with this section and just link to the ancestry.rst\n",
    "    one from the ``seed`` parameter to mutate.\n",
    "```\n",
    "\n",
    "(sec_mutations_discrete)=\n",
    "\n",
    "## Discrete or continuous\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: Examples of using the ``discrete`` argument.\n",
    "```\n",
    "\n",
    "(sec_mutations_time_span)=\n",
    "\n",
    "## Restricting time span\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: Docs and examples for the start_time and end_time params.\n",
    "\n",
    "```\n",
    "\n",
    "(sec_mutations_existing)=\n",
    "\n",
    "## Existing mutations\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: Docs and examples for what to do when you're adding mutations\n",
    "    to a ts that already has mutations. Like what the\n",
    "    ``kept_mutations_before_end_time`` parameter.\n",
    "\n",
    "```\n",
    "\n",
    "(sec_mutations_models)=\n",
    "\n",
    "## Models\n",
    "\n",
    "Mutation models are specified using the `model` parameter to\n",
    "{func}`.mutate`. This parameter can either take the form of a\n",
    "string describing the model (e.g. `model=\"jc69\"`) or an instance of a\n",
    "model definition class (e.g `model=msprime.JC69MutationModel()`).\n",
    "Here are the available models; they are documented in more detail below.\n",
    "\n",
    "- {class}`.BinaryMutationModel`: 0/1 flip-flopping alleles\n",
    "- {class}`.JC69MutationModel`: Jukes & Cantor '69, nucleotides\n",
    "- {class}`.HKYMutationModel`: Hasegawa, Kishino & Yano '85, nucleotides\n",
    "- {class}`.F84MutationModel`: Felsenstein '84, nucleotides\n",
    "- {class}`.GTRMutationModel`: general time-reversible, nucleotides\n",
    "- {class}`.BLOSUM62MutationModel`: amino acids\n",
    "- {class}`.PAMMutationModel`: amino acids\n",
    "- {class}`.MatrixMutationModel`: general finite-state mutations\n",
    "- {class}`.InfiniteAllelesMutationModel`: a generic infinite-alleles model\n",
    "- {class}`.SLiMMutationModel`: infinite-alleles model generating SLiM mutations\n",
    "\n",
    "(sec_mutations_matrix_mutations_models)=\n",
    "\n",
    "### Matrix Mutation Models\n",
    "\n",
    "These classes are defined by an alphabet of possible alleles (`alleles`); an array of\n",
    "probabilities that determines how likely each allele is to be the root, ancestral allele\n",
    "(`root_distribution`); and a transition matrix specifying the probability for each allele\n",
    "to mutate to every other allele. Each class has specific values of these parameters to\n",
    "create the specific model. For your own custom model these parameters can be set using\n",
    "{class}`msprime.MatrixMutationModel`. For more detail about how mutations are simulated\n",
    "in these models see {ref}`sec_mutations_matrix_mutation_models_details`.\n",
    "\n",
    "(sec_mutations_matrix_mutation_models_details)=\n",
    "\n",
    "### Mutation Matrix Models Details\n",
    "\n",
    "Mutation matrix models are specified by three things: an alphabet,\n",
    "a root distribution, and a transition matrix.\n",
    "These leave one free parameter: an overall mutation rate,\n",
    "specified by the mutation `rate` in the call to {func}`.mutate`.\n",
    "Concisely,\n",
    "the underlying model of mutation is a continuous-time Markov chain on the alphabet,\n",
    "started by a draw from `root_distribution`, and\n",
    "with instantaneous transition rate from `i` to `j` that is equal to\n",
    "`rate` multiplied by `transition_matrix[i,j]`.\n",
    "The `root distribution` and every row in the `transition_matrix`\n",
    "must give *probabilities*, i.e., they must be nonnegative numbers summing to 1.\n",
    "For the precise interpretation of these parameters\n",
    "(especially when the transition matrix has nonzero entries on the diagonal)\n",
    "see {ref}`sec_mutations_matrix_mutation_theory`.\n",
    "\n",
    "You can define your own, but you probably don't need to:\n",
    "there are several mutation matrix models already implemented in msprime,\n",
    "using binary (0/1), nucleotide, or amino acid alphabets:\n",
    "\n",
    "### Defining your own finite-sites model\n",
    "\n",
    "If you want to define your own {class}`.MatrixMutationModel`, you have a good\n",
    "deal of freedom. For instance, here's a \"decomposition/growth/disturbance\"\n",
    "mutation model, where the only possible transitions are ðŸŽ„ to ðŸ”¥, ðŸ”¥ to ðŸ’©, and\n",
    "ðŸ’© to ðŸŽ„, with the first transition happening at one-fifth the rate of the\n",
    "other two:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "alleles = [\"ðŸ’©\", \"ðŸŽ„\", \"ðŸ”¥\"]\n",
    "model = msprime.MatrixMutationModel(\n",
    "    alleles,\n",
    "    root_distribution = [1.0, 0.0, 0.0],\n",
    "    transition_matrix = [[0.0, 1.0, 0.0],\n",
    "                         [0.0, 0.8, 0.2],\n",
    "                         [1.0, 0.0, 0.0]]\n",
    ")\n",
    "ts = msprime.simulate(12, Ne=10, random_seed=2, length=7)\n",
    "mts = msprime.mutate(ts, rate=2, random_seed=1, model=model, discrete=True)\n",
    "\n",
    "```\n",
    "\n",
    "We have simulated from this model at rate 2, so the overall rate of mutation\n",
    "from ðŸ’© to ðŸŽ„ and ðŸ”¥ to ðŸ’© is 2, and from ðŸŽ„ to ðŸ”¥ is {math}`2 \\times 0.2\n",
    "= 0.4`. As a result, roughly 5/7th of the states will be ðŸŽ„, with the remainder\n",
    "divided evenly between ðŸ’© and ðŸ”¥. Here is the resulting \"genotype matrix\":\n",
    "\n",
    "```{code-block}\n",
    "\n",
    "for v in mts.variants():\n",
    "   print(\"\".join(v.alleles[k] for k in v.genotypes))\n",
    "\n",
    "ðŸ”¥ðŸŽ„ðŸŽ„ðŸŽ„ðŸ’©ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„\n",
    "ðŸ’©ðŸ’©ðŸ’©ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸ’©ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„\n",
    "ðŸŽ„ðŸŽ„ðŸ”¥ðŸ”¥ðŸ’©ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„\n",
    "ðŸŽ„ðŸŽ„ðŸ’©ðŸ”¥ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸ’©ðŸ’©ðŸ”¥ðŸ’©\n",
    "ðŸ’©ðŸŽ„ðŸŽ„ðŸŽ„ðŸ”¥ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸ”¥ðŸŽ„\n",
    "ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„\n",
    "ðŸ’©ðŸŽ„ðŸŽ„ðŸŽ„ðŸŽ„ðŸ’©ðŸŽ„ðŸŽ„ðŸŽ„ðŸ’©ðŸŽ„ðŸ’©\n",
    "\n",
    "```\n",
    "\n",
    "(sec_mutations_matrix_mutation_theory)=\n",
    "\n",
    "### Parameterization of Matrix Mutation Models\n",
    "\n",
    "Mutation matrix models are specified by three things: an alphabet,\n",
    "a root distribution, and a transition matrix.\n",
    "These leave one free parameter: an overall mutation rate,\n",
    "specified by the mutation `rate` in the call to {func}`.mutate`.\n",
    "Concisely,\n",
    "the underlying model of mutation is a continuous-time Markov chain on the alphabet,\n",
    "started by a draw from `root_distribution`, and\n",
    "with instantaneous transition rate from `i` to `j` that is equal to\n",
    "`rate` multiplied by `transition_matrix[i,j]`.\n",
    "The `root distribution` and every row in the `transition_matrix`\n",
    "must give *probabilities*, i.e., they must be nonnegative numbers summing to 1.\n",
    "\n",
    "To interpret these parameters,\n",
    "it helps to know how the underlying mutational process is implemented.\n",
    "First, \"possible\" mutations are placed on the tree,\n",
    "with a mean density equal to the `rate`, per unit of time and sequence length.\n",
    "If `discrete=False` then this is an infinite-sites model,\n",
    "so each possible mutation occurs at a distinct location.\n",
    "If `discrete=True` then at each integer position,\n",
    "each branch of the tree at that position gets a Poisson number of mutations\n",
    "with mean equal to `rate` multiplied by the length of the branch.\n",
    "Next, each site that has a possible mutation is assigned an ancestral state,\n",
    "i.e., the allele at the root of the tree at that position,\n",
    "by drawing an allele from the probabilities in the `root_distribution`.\n",
    "Now, each possible mutation is examined, moving down the tree.\n",
    "For each, a derived state is chosen using the probabilities given in the\n",
    "row of the `transition_matrix` that corresponds to the \"parental state\",\n",
    "i.e., the allele that this mutation will replace.\n",
    "Importantly, if the chosen allele is the *same* as the parental allele,\n",
    "no mutation is recorded (that's why they were called \"possible mutations\").\n",
    "And, any site at which no mutations are recorded is not recorded either.\n",
    "\n",
    "This arrangement is necessary to fully specify Markov models of mutation,\n",
    "with a free \"mutation rate\" parameter.\n",
    "However, there are some surprising consequences.\n",
    "For instance, the distribution of ancestral alleles, across all sites,\n",
    "is *not* necessarily equal to the root distribution.\n",
    "This is because the root distribution gives the distribution of\n",
    "\"ancestral\" alleles across the entire sequence,\n",
    "but we only see the ancestral alleles at *mutated* sites,\n",
    "and some alleles may have a higher mutation rate than others.\n",
    "For instance, if we have\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "alleles = [\"A\", \"C\", \"G\", \"T\"]\n",
    "root_distribution = np.array([0.25, 0.25, 0.25, 0.25])\n",
    "transition_matrix = np.array([\n",
    "   [0.25, 0.25, 0.25, 0.25],\n",
    "   [ 0.3,  0.0,  0.4,  0.3],\n",
    "   [ 0.3,  0.4,  0.0,  0.3],\n",
    "   [0.25, 0.25, 0.25, 0.25]\n",
    "])\n",
    "\n",
    "```\n",
    "\n",
    "then A and T alleles have a 25% lower mutation rate than do C and G alleles,\n",
    "since 25% of the time that we consider mutating them, we leave them unchanged.\n",
    "From the properties of the Poisson distribution,\n",
    "the probability that a tree of total length {math}`T`\n",
    "has no mutations at a given discrete site is {math}`\\exp(-rT)`,\n",
    "if mutations are put down at a rate of {math}`r`.\n",
    "Suppose that a single tree of total length {math}`T = 1.5`\n",
    "extends over many discrete sites,\n",
    "and that mutations are placed on it at rate {math}`r = 2.0`.\n",
    "Every site that is assigned a \"C\" or \"G\" ancestral allele is retained,\n",
    "but of those sites that are assigned an \"A\" or \"T\",\n",
    "some are not recorded in the resulting tree sequence.\n",
    "The expected proportions of the ancestral states\n",
    "across all sites is proportional to the root distribution\n",
    "multiplied by the probability that at least one mutation is retained on the tree.\n",
    "In this situation it can be computed as follows:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "r = 2.0\n",
    "T = 1.5\n",
    "prob_mut = 1.0 - np.diag(transition_matrix)\n",
    "ancestral_state_distribution = root_distribution * (1 - exp(- r * T * prob_mut))\n",
    "ancestral_state_distribution /= sum(ancestral_state_distribution)\n",
    "\n",
    "```\n",
    "\n",
    "Two more facts about Markov chains are useful to interpret the statistics\n",
    "of these mutation models.\n",
    "First, suppose we have tabulated all mutations, and so for each pair of alleles\n",
    "{math}`i` and {math}`j` we have the proportion of mutations that caused an {math}`i \\to j` change.\n",
    "If allele {math}`i` mutates to a different allele, the chance it mutates to allele {math}`j`\n",
    "is proportional to `transition_matrix[i,j]` but excluding the diagonal (no-change) entry,\n",
    "so is equal to `transition_matrix[i,j] / (1 - transition_matrix[i,i])`.\n",
    "Second, suppose that an ancestor carries allele {math}`i` at a given position.\n",
    "The probability that her descendant some time {math}`t` in the future carries allele {math}`j` at that position\n",
    "is given by a matrix exponential of\n",
    "the scaled [infinitestimal rate matrix](<https://en.wikipedia.org/wiki/Transition_rate_matrix>) of the Markov chain,\n",
    "which can be computed as follows:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "Q = (transition_matrix - np.eye(len(alleles)))\n",
    "Pt = scipy.linalg.expm(t * rate * Q)[i,j]\n",
    "\n",
    "```\n",
    "\n",
    "If the top of a branch of length {math}`t` has allele {math}`i`,\n",
    "the bottom of the branch has allele {math}`j` with probability `Pt[i,j]`.\n",
    "\n",
    "(sec_mutations_mutation_infinite_alleles)=\n",
    "\n",
    "### Infinite Alleles Mutation Models\n",
    "\n",
    "You can also use a model of *infinite alleles* mutation: where each new mutation produces a unique,\n",
    "never-before-seen allele. The underlying mutation model just assigns the derived state\n",
    "to be a new integer every time a new mutation appears.\n",
    "By default these integers start at zero, but a different starting point can be chosen,\n",
    "with the `start_allele` argument.\n",
    "It does this globally across all mutations, so that the first assigned allele will be `start_allele`,\n",
    "and if `n` alleles are assigned in total (across ancestral and derived states),\n",
    "these will be the next `n-1` integers.\n",
    "Many theoretical results are derived based on this mutation model (e.g., Ewens' sampling formula).\n",
    "\n",
    "For instance, here we'll simulate with the infinite alleles model on a single tree,\n",
    "and print the resulting tree, labeling each mutation with its derived state:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "ts = msprime.simulate(12, random_seed=2, length=1)\n",
    "model = msprime.InfiniteAllelesMutationModel()\n",
    "mts = msprime.mutate(ts, rate=2, random_seed=1, model=model, discrete=True)\n",
    "t = mts.first()\n",
    "ml = {m.id: m.derived_state for m in mts.mutations()}\n",
    "t.draw_svg('infinite_alleles.svg', mutation_labels=ml, node_labels={}, size=(400, 300))\n",
    "\n",
    "```\n",
    "\n",
    "Apparently, there were 20 mutations at this site, but the alleles present in the population are\n",
    "\"13\" (in five copies), \"17\" (in two copies), and one copy each of \"14\", \"15\", \"19\", and \"20\".\n",
    "\n",
    "```{image} _static/infinite_alleles.svg\n",
    "\n",
    "```\n",
    "\n",
    ":::{warning}\n",
    "\n",
    "Neither this nor the next infinite alleles mutation model check to see if the alleles\n",
    "they produce already exist at the mutated sites. So, if you are using these\n",
    "models to add mutations to an already-mutated tree sequence, it is up to you\n",
    "to set the starting allele appropriately, and to make sure the results make sense!\n",
    "\n",
    ":::\n",
    "\n",
    "(sec_mutations_mutation_slim_mutations)=\n",
    "\n",
    "### SLiM mutations\n",
    "\n",
    "A special class of infinite alleles model is provided for use with [SLiM](<https://messerlab.org/slim/>),\n",
    "to agree with the underlying mutation model in SLiM.\n",
    "As with the InfiniteAlleles model, it assigns each new mutation a unique integer,\n",
    "by keeping track of the `next_id` and incrementing it each time a new mutation appears.\n",
    "\n",
    "This differs from the {class}`.InfiniteAllelesMutationmodel` because mutations\n",
    "in SLiM can \"stack\": new mutations can add to the existing state, rather than\n",
    "replacing the previous state. So, derived states are comma-separated lists of\n",
    "mutation IDs, and the ancestral state is always the empty string. For instance,\n",
    "if a new mutation with ID 5 occurs at a site, and then later another mutation\n",
    "appears with ID 64, the sequence of alleles moving along this line of descent\n",
    "would be `\"\"`, then `\"5\"`, and finally `\"5,64\"`. Furthermore, the mutation\n",
    "model adds SLiM metadata to each mutation, which records, among other things,\n",
    "the SLiM mutation type of each mutation, and the selection coefficient (which\n",
    "is always 0.0, since adding mutations in this way only makes sense if they are\n",
    "neutral). For this reason, the model has one required parameter: the `type`\n",
    "of the mutation, a nonnegative integer. If, for instance, you specify\n",
    "`type=1`, then the mutations in SLiM will be of type `m1`. For more\n",
    "information, and for how to modify the metadata (e.g., changing the selection\n",
    "coefficients), see\n",
    "[the pyslim documentation](<https://pyslim.readthedocs.io/en/latest/>).\n",
    "For instance,\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "model = msprime.SLiMMutationModel(type=1)\n",
    "mts = msprime.mutate(ts, rate=1, random_seed=1, model=model, discrete=True)\n",
    "t = mts.first()\n",
    "ml = {m.id: m.derived_state for m in mts.mutations()}\n",
    "t.draw_svg('slim_alleles.svg', mutation_labels=ml, node_labels={}, size=(400, 300))\n",
    "\n",
    "```\n",
    "\n",
    "The resulting alleles show how derived states are built:\n",
    "\n",
    "```{image} _static/slim_alleles.svg\n",
    "\n",
    "```\n",
    "\n",
    "The behavior of this mutation model when used to add mutations to a previously mutated\n",
    "tree sequence can be subtle. Let's look at a simple example.\n",
    "Here, we first lay down mutations of type 1, starting from ID 0:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "model_1 = msprime.SLiMMutationModel(type=1)\n",
    "mts_1 = msprime.mutate(ts, rate=0.5, random_seed=2, model=model_1, discrete=True)\n",
    "t = mts_1.first()\n",
    "ml = {m.id: m.derived_state for m in mts_1.mutations()}\n",
    "t.draw_svg(f'slim_alleles_1.svg', mutation_labels=ml, node_labels={}, size=(400, 300))\n",
    "\n",
    "```\n",
    "\n",
    "```{image} _static/slim_alleles_1.svg\n",
    "\n",
    "```\n",
    "\n",
    "Next, we lay down mutations of type 2.\n",
    "These we assign starting from ID 100,\n",
    "to make it easy to see which are which:\n",
    "in general just need to make sure that we start at an ID greater than any\n",
    "previously assigned.\n",
    "Note the `allow_ancestral=True` parameter:\n",
    "this would error, otherwise because we are adding mutations above existing ones.\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "model_2 = msprime.SLiMMutationModel(type=2, next_id=100)\n",
    "mts = msprime.mutate(mts_1, rate=0.5, random_seed=3, model=model_2, discrete=True,\n",
    "                     keep=True, allow_ancestral=True)\n",
    "t = mts.first()\n",
    "ml = {m.id: m.derived_state for m in mts.mutations()}\n",
    "t.draw_svg(f'slim_alleles_12.svg', mutation_labels=ml, node_labels={}, size=(400, 300))\n",
    "\n",
    "```\n",
    "\n",
    "```{image} _static/slim_alleles_12.svg\n",
    "\n",
    "```\n",
    "\n",
    "Note what has happened here: on the top branch on the right side of the tree,\n",
    "with the first model we added two mutations: first a mutation with ID `0`,\n",
    "then a mutation with ID `3`.\n",
    "Then, with the second model, we added two more mutations to this same branch,\n",
    "with IDs `100` and `102`, between these two mutations.\n",
    "These were added to mutation `0`, obtaining alleles `0,100` and `0,100,102`.\n",
    "But then, moving down the branch, we come upon the mutation with ID `3`.\n",
    "This was already present in the tree sequence, so its derived state is not modified:\n",
    "`0,3`. We can rationalize this, post-hoc, by saying that the type 1 mutation `3`\n",
    "has \"erased\" the type 2 mutations `100` and `102`.\n",
    "If you want a different arrangment,\n",
    "you can go back and edit the derived states (and metadata) as you like."
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.12,
    "jupytext_version": "1.9.1"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.7"
  },
  "source_map": [
   14,
   22
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}